// @firebase/webchannel-wrapper/dist/esm/index.esm2017.jsx
var Bn = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, h, bt = bt || {}, l = Bn || self;
function jt(t) {
  var i = typeof t;
  return i = i != "object" ? i : t ? Array.isArray(t) ? "array" : i : "null", i == "array" || i == "object" && typeof t.length == "number";
}
function rt(t) {
  var i = typeof t;
  return i == "object" && t != null || i == "function";
}
function Fn(t) {
  return Object.prototype.hasOwnProperty.call(t, Nt) && t[Nt] || (t[Nt] = ++Un);
}
var Nt = "closure_uid_" + (1e9 * Math.random() >>> 0), Un = 0;
function Kn(t, i, n) {
  return t.call.apply(t.bind, arguments);
}
function Gn(t, i, n) {
  if (!t)
    throw Error();
  if (2 < arguments.length) {
    var e = Array.prototype.slice.call(arguments, 2);
    return function() {
      var r = Array.prototype.slice.call(arguments);
      return Array.prototype.unshift.apply(r, e), t.apply(i, r);
    };
  }
  return function() {
    return t.apply(i, arguments);
  };
}
function T(t, i, n) {
  return Function.prototype.bind && Function.prototype.bind.toString().indexOf("native code") != -1 ? T = Kn : T = Gn, T.apply(null, arguments);
}
function at(t, i) {
  var n = Array.prototype.slice.call(arguments, 1);
  return function() {
    var e = n.slice();
    return e.push.apply(e, arguments), t.apply(this, e);
  };
}
function d(t, i) {
  function n() {
  }
  n.prototype = i.prototype, t.$ = i.prototype, t.prototype = new n(), t.prototype.constructor = t, t.ac = function(e, r, s) {
    for (var o = Array(arguments.length - 2), f = 2; f < arguments.length; f++)
      o[f - 2] = arguments[f];
    return i.prototype[r].apply(e, o);
  };
}
function k() {
  this.s = this.s, this.o = this.o;
}
var Jn = 0;
k.prototype.s = !1;
k.prototype.sa = function() {
  !this.s && (this.s = !0, this.N(), Jn != 0) && Fn(this);
};
k.prototype.N = function() {
  if (this.o)
    for (; this.o.length; )
      this.o.shift()();
};
const Xi = Array.prototype.indexOf ? function(t, i) {
  return Array.prototype.indexOf.call(t, i, void 0);
} : function(t, i) {
  if (typeof t == "string")
    return typeof i != "string" || i.length != 1 ? -1 : t.indexOf(i, 0);
  for (let n = 0; n < t.length; n++)
    if (n in t && t[n] === i)
      return n;
  return -1;
};
function ti(t) {
  const i = t.length;
  if (0 < i) {
    const n = Array(i);
    for (let e = 0; e < i; e++)
      n[e] = t[e];
    return n;
  }
  return [];
}
function ji(t, i) {
  for (let n = 1; n < arguments.length; n++) {
    const e = arguments[n];
    if (jt(e)) {
      const r = t.length || 0, s = e.length || 0;
      t.length = r + s;
      for (let o = 0; o < s; o++)
        t[r + o] = e[o];
    } else
      t.push(e);
  }
}
function E(t, i) {
  this.type = t, this.g = this.target = i, this.defaultPrevented = !1;
}
E.prototype.h = function() {
  this.defaultPrevented = !0;
};
var Yn = function() {
  if (!l.addEventListener || !Object.defineProperty)
    return !1;
  var t = !1, i = Object.defineProperty({}, "passive", { get: function() {
    t = !0;
  } });
  try {
    const n = () => {
    };
    l.addEventListener("test", n, i), l.removeEventListener("test", n, i);
  } catch {
  }
  return t;
}();
function W(t) {
  return /^[\s\xa0]*$/.test(t);
}
function At() {
  var t = l.navigator;
  return t && (t = t.userAgent) ? t : "";
}
function C(t) {
  return At().indexOf(t) != -1;
}
function ii(t) {
  return ii[" "](t), t;
}
ii[" "] = function() {
};
function zn(t, i) {
  var n = Be;
  return Object.prototype.hasOwnProperty.call(n, t) ? n[t] : n[t] = i(t);
}
var qn = C("Opera"), $ = C("Trident") || C("MSIE"), Ni = C("Edge"), Gt = Ni || $, Bi = C("Gecko") && !(At().toLowerCase().indexOf("webkit") != -1 && !C("Edge")) && !(C("Trident") || C("MSIE")) && !C("Edge"), Vn = At().toLowerCase().indexOf("webkit") != -1 && !C("Edge");
function Fi() {
  var t = l.document;
  return t ? t.documentMode : void 0;
}
t: {
  var Ai = "", Bt = function() {
    var t = At();
    if (Bi)
      return /rv:([^\);]+)(\)|;)/.exec(t);
    if (Ni)
      return /Edge\/([\d\.]+)/.exec(t);
    if ($)
      return /\b(?:MSIE|rv)[: ]([^\);]+)(\)|;)/.exec(t);
    if (Vn)
      return /WebKit\/(\S+)/.exec(t);
    if (qn)
      return /(?:Version)[ \/]?(\S+)/.exec(t);
  }();
  if (Bt && (Ai = Bt ? Bt[1] : ""), $) {
    var Ci = Fi();
    if (Ci != null && Ci > parseFloat(Ai))
      break t;
  }
}
l.document && $ && Fi();
function Z(t, i) {
  if (E.call(this, t ? t.type : ""), this.relatedTarget = this.g = this.target = null, this.button = this.screenY = this.screenX = this.clientY = this.clientX = 0, this.key = "", this.metaKey = this.shiftKey = this.altKey = this.ctrlKey = !1, this.state = null, this.pointerId = 0, this.pointerType = "", this.i = null, t) {
    var n = this.type = t.type, e = t.changedTouches && t.changedTouches.length ? t.changedTouches[0] : null;
    if (this.target = t.target || t.srcElement, this.g = i, i = t.relatedTarget) {
      if (Bi) {
        t: {
          try {
            ii(i.nodeName);
            var r = !0;
            break t;
          } catch {
          }
          r = !1;
        }
        r || (i = null);
      }
    } else
      n == "mouseover" ? i = t.fromElement : n == "mouseout" && (i = t.toElement);
    this.relatedTarget = i, e ? (this.clientX = e.clientX !== void 0 ? e.clientX : e.pageX, this.clientY = e.clientY !== void 0 ? e.clientY : e.pageY, this.screenX = e.screenX || 0, this.screenY = e.screenY || 0) : (this.clientX = t.clientX !== void 0 ? t.clientX : t.pageX, this.clientY = t.clientY !== void 0 ? t.clientY : t.pageY, this.screenX = t.screenX || 0, this.screenY = t.screenY || 0), this.button = t.button, this.key = t.key || "", this.ctrlKey = t.ctrlKey, this.altKey = t.altKey, this.shiftKey = t.shiftKey, this.metaKey = t.metaKey, this.pointerId = t.pointerId || 0, this.pointerType = typeof t.pointerType == "string" ? t.pointerType : Wn[t.pointerType] || "", this.state = t.state, this.i = t, t.defaultPrevented && Z.$.h.call(this);
  }
}
d(Z, E);
var Wn = { 2: "touch", 3: "pen", 4: "mouse" };
Z.prototype.h = function() {
  Z.$.h.call(this);
  var t = this.i;
  t.preventDefault ? t.preventDefault() : t.returnValue = !1;
};
var st = "closure_listenable_" + (1e6 * Math.random() | 0), $n = 0;
function Zn(t, i, n, e, r) {
  this.listener = t, this.proxy = null, this.src = i, this.type = n, this.capture = !!e, this.la = r, this.key = ++$n, this.fa = this.ia = !1;
}
function Ct(t) {
  t.fa = !0, t.listener = null, t.proxy = null, t.src = null, t.la = null;
}
function ni(t, i, n) {
  for (const e in t)
    i.call(n, t[e], e, t);
}
function Qn(t, i) {
  for (const n in t)
    i.call(void 0, t[n], n, t);
}
function Ui(t) {
  const i = {};
  for (const n in t)
    i[n] = t[n];
  return i;
}
const xi = "constructor hasOwnProperty isPrototypeOf propertyIsEnumerable toLocaleString toString valueOf".split(" ");
function Ki(t, i) {
  let n, e;
  for (let r = 1; r < arguments.length; r++) {
    e = arguments[r];
    for (n in e)
      t[n] = e[n];
    for (let s = 0; s < xi.length; s++)
      n = xi[s], Object.prototype.hasOwnProperty.call(e, n) && (t[n] = e[n]);
  }
}
function xt(t) {
  this.src = t, this.g = {}, this.h = 0;
}
xt.prototype.add = function(t, i, n, e, r) {
  var s = t.toString();
  t = this.g[s], t || (t = this.g[s] = [], this.h++);
  var o = Yt(t, i, e, r);
  return -1 < o ? (i = t[o], n || (i.ia = !1)) : (i = new Zn(i, this.src, s, !!e, r), i.ia = n, t.push(i)), i;
};
function Jt(t, i) {
  var n = i.type;
  if (n in t.g) {
    var e = t.g[n], r = Xi(e, i), s;
    (s = 0 <= r) && Array.prototype.splice.call(e, r, 1), s && (Ct(i), t.g[n].length == 0 && (delete t.g[n], t.h--));
  }
}
function Yt(t, i, n, e) {
  for (var r = 0; r < t.length; ++r) {
    var s = t[r];
    if (!s.fa && s.listener == i && s.capture == !!n && s.la == e)
      return r;
  }
  return -1;
}
var ei = "closure_lm_" + (1e6 * Math.random() | 0), Ft = {};
function Gi(t, i, n, e, r) {
  if (e && e.once)
    return Yi(t, i, n, e, r);
  if (Array.isArray(i)) {
    for (var s = 0; s < i.length; s++)
      Gi(t, i[s], n, e, r);
    return null;
  }
  return n = oi(n), t && t[st] ? t.O(i, n, rt(e) ? !!e.capture : !!e, r) : Ji(t, i, n, !1, e, r);
}
function Ji(t, i, n, e, r, s) {
  if (!i)
    throw Error("Invalid event type");
  var o = rt(r) ? !!r.capture : !!r, f = si(t);
  if (f || (t[ei] = f = new xt(t)), n = f.add(i, n, e, o, s), n.proxy)
    return n;
  if (e = bn(), n.proxy = e, e.src = t, e.listener = n, t.addEventListener)
    Yn || (r = o), r === void 0 && (r = !1), t.addEventListener(i.toString(), e, r);
  else if (t.attachEvent)
    t.attachEvent(qi(i.toString()), e);
  else if (t.addListener && t.removeListener)
    t.addListener(e);
  else
    throw Error("addEventListener and attachEvent are unavailable.");
  return n;
}
function bn() {
  function t(n) {
    return i.call(t.src, t.listener, n);
  }
  const i = te;
  return t;
}
function Yi(t, i, n, e, r) {
  if (Array.isArray(i)) {
    for (var s = 0; s < i.length; s++)
      Yi(t, i[s], n, e, r);
    return null;
  }
  return n = oi(n), t && t[st] ? t.P(i, n, rt(e) ? !!e.capture : !!e, r) : Ji(t, i, n, !0, e, r);
}
function zi(t, i, n, e, r) {
  if (Array.isArray(i))
    for (var s = 0; s < i.length; s++)
      zi(t, i[s], n, e, r);
  else
    e = rt(e) ? !!e.capture : !!e, n = oi(n), t && t[st] ? (t = t.i, i = String(i).toString(), i in t.g && (s = t.g[i], n = Yt(s, n, e, r), -1 < n && (Ct(s[n]), Array.prototype.splice.call(s, n, 1), s.length == 0 && (delete t.g[i], t.h--)))) : t && (t = si(t)) && (i = t.g[i.toString()], t = -1, i && (t = Yt(i, n, e, r)), (n = -1 < t ? i[t] : null) && ri(n));
}
function ri(t) {
  if (typeof t != "number" && t && !t.fa) {
    var i = t.src;
    if (i && i[st])
      Jt(i.i, t);
    else {
      var n = t.type, e = t.proxy;
      i.removeEventListener ? i.removeEventListener(n, e, t.capture) : i.detachEvent ? i.detachEvent(qi(n), e) : i.addListener && i.removeListener && i.removeListener(e), (n = si(i)) ? (Jt(n, t), n.h == 0 && (n.src = null, i[ei] = null)) : Ct(t);
    }
  }
}
function qi(t) {
  return t in Ft ? Ft[t] : Ft[t] = "on" + t;
}
function te(t, i) {
  if (t.fa)
    t = !0;
  else {
    i = new Z(i, this);
    var n = t.listener, e = t.la || t.src;
    t.ia && ri(t), t = n.call(e, i);
  }
  return t;
}
function si(t) {
  return t = t[ei], t instanceof xt ? t : null;
}
var Ut = "__closure_events_fn_" + (1e9 * Math.random() >>> 0);
function oi(t) {
  return typeof t == "function" ? t : (t[Ut] || (t[Ut] = function(i) {
    return t.handleEvent(i);
  }), t[Ut]);
}
function v() {
  k.call(this), this.i = new xt(this), this.S = this, this.J = null;
}
d(v, k);
v.prototype[st] = !0;
v.prototype.removeEventListener = function(t, i, n, e) {
  zi(this, t, i, n, e);
};
function w(t, i) {
  var n, e = t.J;
  if (e)
    for (n = []; e; e = e.J)
      n.push(e);
  if (t = t.S, e = i.type || i, typeof i == "string")
    i = new E(i, t);
  else if (i instanceof E)
    i.target = i.target || t;
  else {
    var r = i;
    i = new E(e, t), Ki(i, r);
  }
  if (r = !0, n)
    for (var s = n.length - 1; 0 <= s; s--) {
      var o = i.g = n[s];
      r = yt(o, e, !0, i) && r;
    }
  if (o = i.g = t, r = yt(o, e, !0, i) && r, r = yt(o, e, !1, i) && r, n)
    for (s = 0; s < n.length; s++)
      o = i.g = n[s], r = yt(o, e, !1, i) && r;
}
v.prototype.N = function() {
  if (v.$.N.call(this), this.i) {
    var t = this.i, i;
    for (i in t.g) {
      for (var n = t.g[i], e = 0; e < n.length; e++)
        Ct(n[e]);
      delete t.g[i], t.h--;
    }
  }
  this.J = null;
};
v.prototype.O = function(t, i, n, e) {
  return this.i.add(String(t), i, !1, n, e);
};
v.prototype.P = function(t, i, n, e) {
  return this.i.add(String(t), i, !0, n, e);
};
function yt(t, i, n, e) {
  if (i = t.i.g[String(i)], !i)
    return !0;
  i = i.concat();
  for (var r = !0, s = 0; s < i.length; ++s) {
    var o = i[s];
    if (o && !o.fa && o.capture == n) {
      var f = o.listener, a = o.la || o.src;
      o.ia && Jt(t.i, o), r = f.call(a, e) !== !1 && r;
    }
  }
  return r && !e.defaultPrevented;
}
var hi = l.JSON.stringify;
class ie {
  constructor(i, n) {
    this.i = i, this.j = n, this.h = 0, this.g = null;
  }
  get() {
    let i;
    return 0 < this.h ? (this.h--, i = this.g, this.g = i.next, i.next = null) : i = this.i(), i;
  }
}
function ne() {
  var t = li;
  let i = null;
  return t.g && (i = t.g, t.g = t.g.next, t.g || (t.h = null), i.next = null), i;
}
class ee {
  constructor() {
    this.h = this.g = null;
  }
  add(i, n) {
    const e = Vi.get();
    e.set(i, n), this.h ? this.h.next = e : this.g = e, this.h = e;
  }
}
var Vi = new ie(() => new re(), (t) => t.reset());
class re {
  constructor() {
    this.next = this.g = this.h = null;
  }
  set(i, n) {
    this.h = i, this.g = n, this.next = null;
  }
  reset() {
    this.next = this.g = this.h = null;
  }
}
function se(t) {
  var i = 1;
  t = t.split(":");
  const n = [];
  for (; 0 < i && t.length; )
    n.push(t.shift()), i--;
  return t.length && n.push(t.join(":")), n;
}
function oe(t) {
  l.setTimeout(() => {
    throw t;
  }, 0);
}
let Q, b = !1, li = new ee(), Wi = () => {
  const t = l.Promise.resolve(void 0);
  Q = () => {
    t.then(he);
  };
};
var he = () => {
  for (var t; t = ne(); ) {
    try {
      t.h.call(t.g);
    } catch (n) {
      oe(n);
    }
    var i = Vi;
    i.j(t), 100 > i.h && (i.h++, t.next = i.g, i.g = t);
  }
  b = !1;
};
function It(t, i) {
  v.call(this), this.h = t || 1, this.g = i || l, this.j = T(this.qb, this), this.l = Date.now();
}
d(It, v);
h = It.prototype;
h.ga = !1;
h.T = null;
h.qb = function() {
  if (this.ga) {
    var t = Date.now() - this.l;
    0 < t && t < 0.8 * this.h ? this.T = this.g.setTimeout(this.j, this.h - t) : (this.T && (this.g.clearTimeout(this.T), this.T = null), w(this, "tick"), this.ga && (ui(this), this.start()));
  }
};
h.start = function() {
  this.ga = !0, this.T || (this.T = this.g.setTimeout(this.j, this.h), this.l = Date.now());
};
function ui(t) {
  t.ga = !1, t.T && (t.g.clearTimeout(t.T), t.T = null);
}
h.N = function() {
  It.$.N.call(this), ui(this), delete this.g;
};
function fi(t, i, n) {
  if (typeof t == "function")
    n && (t = T(t, n));
  else if (t && typeof t.handleEvent == "function")
    t = T(t.handleEvent, t);
  else
    throw Error("Invalid listener argument");
  return 2147483647 < Number(i) ? -1 : l.setTimeout(t, i || 0);
}
function $i(t) {
  t.g = fi(() => {
    t.g = null, t.i && (t.i = !1, $i(t));
  }, t.j);
  const i = t.h;
  t.h = null, t.m.apply(null, i);
}
class le extends k {
  constructor(i, n) {
    super(), this.m = i, this.j = n, this.h = null, this.i = !1, this.g = null;
  }
  l(i) {
    this.h = arguments, this.g ? this.i = !0 : $i(this);
  }
  N() {
    super.N(), this.g && (l.clearTimeout(this.g), this.g = null, this.i = !1, this.h = null);
  }
}
function tt(t) {
  k.call(this), this.h = t, this.g = {};
}
d(tt, k);
var Ii = [];
function Zi(t, i, n, e) {
  Array.isArray(n) || (n && (Ii[0] = n.toString()), n = Ii);
  for (var r = 0; r < n.length; r++) {
    var s = Gi(i, n[r], e || t.handleEvent, !1, t.h || t);
    if (!s)
      break;
    t.g[s.key] = s;
  }
}
function Qi(t) {
  ni(t.g, function(i, n) {
    this.g.hasOwnProperty(n) && ri(i);
  }, t), t.g = {};
}
tt.prototype.N = function() {
  tt.$.N.call(this), Qi(this);
};
tt.prototype.handleEvent = function() {
  throw Error("EventHandler.handleEvent not implemented");
};
function Mt() {
  this.g = !0;
}
Mt.prototype.Ea = function() {
  this.g = !1;
};
function ue(t, i, n, e, r, s) {
  t.info(function() {
    if (t.g)
      if (s)
        for (var o = "", f = s.split("&"), a = 0; a < f.length; a++) {
          var u = f[a].split("=");
          if (1 < u.length) {
            var y = u[0];
            u = u[1];
            var P = y.split("_");
            o = 2 <= P.length && P[1] == "type" ? o + (y + "=" + u + "&") : o + (y + "=redacted&");
          }
        }
      else
        o = null;
    else
      o = s;
    return "XMLHTTP REQ (" + e + ") [attempt " + r + "]: " + i + `
` + n + `
` + o;
  });
}
function fe(t, i, n, e, r, s, o) {
  t.info(function() {
    return "XMLHTTP RESP (" + e + ") [ attempt " + r + "]: " + i + `
` + n + `
` + s + " " + o;
  });
}
function N(t, i, n, e) {
  t.info(function() {
    return "XMLHTTP TEXT (" + i + "): " + pe(t, n) + (e ? " " + e : "");
  });
}
function ge(t, i) {
  t.info(function() {
    return "TIMEOUT: " + i;
  });
}
Mt.prototype.info = function() {
};
function pe(t, i) {
  if (!t.g)
    return i;
  if (!i)
    return null;
  try {
    var n = JSON.parse(i);
    if (n) {
      for (t = 0; t < n.length; t++)
        if (Array.isArray(n[t])) {
          var e = n[t];
          if (!(2 > e.length)) {
            var r = e[1];
            if (Array.isArray(r) && !(1 > r.length)) {
              var s = r[0];
              if (s != "noop" && s != "stop" && s != "close")
                for (var o = 1; o < r.length; o++)
                  r[o] = "";
            }
          }
        }
    }
    return hi(n);
  } catch {
    return i;
  }
}
var F = {}, Mi = null;
function gi() {
  return Mi = Mi || new v();
}
F.Ta = "serverreachability";
function bi(t) {
  E.call(this, F.Ta, t);
}
d(bi, E);
function it(t) {
  const i = gi();
  w(i, new bi(i));
}
F.STAT_EVENT = "statevent";
function tn(t, i) {
  E.call(this, F.STAT_EVENT, t), this.stat = i;
}
d(tn, E);
function S(t) {
  const i = gi();
  w(i, new tn(i, t));
}
F.Ua = "timingevent";
function nn(t, i) {
  E.call(this, F.Ua, t), this.size = i;
}
d(nn, E);
function ot(t, i) {
  if (typeof t != "function")
    throw Error("Fn must not be null and must be a function");
  return l.setTimeout(function() {
    t();
  }, i);
}
var pi = { NO_ERROR: 0, rb: 1, Eb: 2, Db: 3, yb: 4, Cb: 5, Fb: 6, Qa: 7, TIMEOUT: 8, Ib: 9 }, ce = { wb: "complete", Sb: "success", Ra: "error", Qa: "abort", Kb: "ready", Lb: "readystatechange", TIMEOUT: "timeout", Gb: "incrementaldata", Jb: "progress", zb: "downloadprogress", $b: "uploadprogress" };
function ci() {
}
ci.prototype.h = null;
function Ri(t) {
  return t.h || (t.h = t.i());
}
function ae() {
}
var ht = { OPEN: "a", vb: "b", Ra: "c", Hb: "d" };
function ai() {
  E.call(this, "d");
}
d(ai, E);
function yi() {
  E.call(this, "c");
}
d(yi, E);
var zt;
function Rt() {
}
d(Rt, ci);
Rt.prototype.g = function() {
  return new XMLHttpRequest();
};
Rt.prototype.i = function() {
  return {};
};
zt = new Rt();
function lt(t, i, n, e) {
  this.l = t, this.j = i, this.m = n, this.W = e || 1, this.U = new tt(this), this.P = ye, t = Gt ? 125 : void 0, this.V = new It(t), this.I = null, this.i = !1, this.u = this.B = this.A = this.L = this.G = this.Y = this.C = null, this.F = [], this.g = null, this.o = 0, this.s = this.v = null, this.ca = -1, this.J = !1, this.O = 0, this.M = null, this.ba = this.K = this.aa = this.S = !1, this.h = new en();
}
function en() {
  this.i = null, this.g = "", this.h = !1;
}
var ye = 45e3, rn = {}, qt = {};
h = lt.prototype;
h.setTimeout = function(t) {
  this.P = t;
};
function Vt(t, i, n) {
  t.L = 1, t.A = Pt(O(i)), t.u = n, t.S = !0, sn(t, null);
}
function sn(t, i) {
  t.G = Date.now(), ut(t), t.B = O(t.A);
  var n = t.B, e = t.W;
  Array.isArray(e) || (e = [String(e)]), cn(n.i, "t", e), t.o = 0, n = t.l.J, t.h = new en(), t.g = Ln(t.l, n ? i : null, !t.u), 0 < t.O && (t.M = new le(T(t.Pa, t, t.g), t.O)), Zi(t.U, t.g, "readystatechange", t.nb), i = t.I ? Ui(t.I) : {}, t.u ? (t.v || (t.v = "POST"), i["Content-Type"] = "application/x-www-form-urlencoded", t.g.ha(t.B, t.v, t.u, i)) : (t.v = "GET", t.g.ha(t.B, t.v, null, i)), it(), ue(t.j, t.v, t.B, t.m, t.W, t.u);
}
h.nb = function(t) {
  t = t.target;
  const i = this.M;
  i && x(t) == 3 ? i.l() : this.Pa(t);
};
h.Pa = function(t) {
  try {
    if (t == this.g)
      t: {
        const y = x(this.g);
        var i = this.g.Ia();
        const P = this.g.da();
        if (!(3 > y) && (y != 3 || Gt || this.g && (this.h.h || this.g.ja() || Di(this.g)))) {
          this.J || y != 4 || i == 7 || (i == 8 || 0 >= P ? it(3) : it(2)), Ot(this);
          var n = this.g.da();
          this.ca = n;
          i:
            if (on(this)) {
              var e = Di(this.g);
              t = "";
              var r = e.length, s = x(this.g) == 4;
              if (!this.h.i) {
                if (typeof TextDecoder > "u") {
                  L(this), V(this);
                  var o = "";
                  break i;
                }
                this.h.i = new l.TextDecoder();
              }
              for (i = 0; i < r; i++)
                this.h.h = !0, t += this.h.i.decode(e[i], { stream: s && i == r - 1 });
              e.length = 0, this.h.g += t, this.o = 0, o = this.h.g;
            } else
              o = this.g.ja();
          if (this.i = n == 200, fe(this.j, this.v, this.B, this.m, this.W, y, n), this.i) {
            if (this.aa && !this.K) {
              i: {
                if (this.g) {
                  var f, a = this.g;
                  if ((f = a.g ? a.g.getResponseHeader("X-HTTP-Initial-Response") : null) && !W(f)) {
                    var u = f;
                    break i;
                  }
                }
                u = null;
              }
              if (n = u)
                N(this.j, this.m, n, "Initial handshake response via X-HTTP-Initial-Response"), this.K = !0, Wt(this, n);
              else {
                this.i = !1, this.s = 3, S(12), L(this), V(this);
                break t;
              }
            }
            this.S ? (hn(this, y, o), Gt && this.i && y == 3 && (Zi(this.U, this.V, "tick", this.mb), this.V.start())) : (N(this.j, this.m, o, null), Wt(this, o)), y == 4 && L(this), this.i && !this.J && (y == 4 ? On(this.l, this) : (this.i = !1, ut(this)));
          } else
            He(this.g), n == 400 && 0 < o.indexOf("Unknown SID") ? (this.s = 3, S(12)) : (this.s = 0, S(13)), L(this), V(this);
        }
      }
  } catch {
  } finally {
  }
};
function on(t) {
  return t.g ? t.v == "GET" && t.L != 2 && t.l.Ha : !1;
}
function hn(t, i, n) {
  let e = !0, r;
  for (; !t.J && t.o < n.length; )
    if (r = ve(t, n), r == qt) {
      i == 4 && (t.s = 4, S(14), e = !1), N(t.j, t.m, null, "[Incomplete Response]");
      break;
    } else if (r == rn) {
      t.s = 4, S(15), N(t.j, t.m, n, "[Invalid Chunk]"), e = !1;
      break;
    } else
      N(t.j, t.m, r, null), Wt(t, r);
  on(t) && t.o != 0 && (t.h.g = t.h.g.slice(t.o), t.o = 0), i != 4 || n.length != 0 || t.h.h || (t.s = 1, S(16), e = !1), t.i = t.i && e, e ? 0 < n.length && !t.ba && (t.ba = !0, i = t.l, i.g == t && i.ca && !i.M && (i.l.info("Great, no buffering proxy detected. Bytes received: " + n.length), Ei(i), i.M = !0, S(11))) : (N(
    t.j,
    t.m,
    n,
    "[Invalid Chunked Response]"
  ), L(t), V(t));
}
h.mb = function() {
  if (this.g) {
    var t = x(this.g), i = this.g.ja();
    this.o < i.length && (Ot(this), hn(this, t, i), this.i && t != 4 && ut(this));
  }
};
function ve(t, i) {
  var n = t.o, e = i.indexOf(`
`, n);
  return e == -1 ? qt : (n = Number(i.substring(n, e)), isNaN(n) ? rn : (e += 1, e + n > i.length ? qt : (i = i.slice(e, e + n), t.o = e + n, i)));
}
h.cancel = function() {
  this.J = !0, L(this);
};
function ut(t) {
  t.Y = Date.now() + t.P, ln(t, t.P);
}
function ln(t, i) {
  if (t.C != null)
    throw Error("WatchDog timer not null");
  t.C = ot(T(t.lb, t), i);
}
function Ot(t) {
  t.C && (l.clearTimeout(t.C), t.C = null);
}
h.lb = function() {
  this.C = null;
  const t = Date.now();
  0 <= t - this.Y ? (ge(this.j, this.B), this.L != 2 && (it(), S(17)), L(this), this.s = 2, V(this)) : ln(this, this.Y - t);
};
function V(t) {
  t.l.H == 0 || t.J || On(t.l, t);
}
function L(t) {
  Ot(t);
  var i = t.M;
  i && typeof i.sa == "function" && i.sa(), t.M = null, ui(t.V), Qi(t.U), t.g && (i = t.g, t.g = null, i.abort(), i.sa());
}
function Wt(t, i) {
  try {
    var n = t.l;
    if (n.H != 0 && (n.g == t || $t(n.i, t))) {
      if (!t.K && $t(n.i, t) && n.H == 3) {
        try {
          var e = n.Ja.g.parse(i);
        } catch {
          e = null;
        }
        if (Array.isArray(e) && e.length == 3) {
          var r = e;
          if (r[0] == 0) {
            t:
              if (!n.u) {
                if (n.g)
                  if (n.g.G + 3e3 < t.G)
                    Tt(n), _t(n);
                  else
                    break t;
                Ti(n), S(18);
              }
          } else
            n.Fa = r[1], 0 < n.Fa - n.V && 37500 > r[2] && n.G && n.A == 0 && !n.v && (n.v = ot(T(n.ib, n), 6e3));
          if (1 >= vn(n.i) && n.oa) {
            try {
              n.oa();
            } catch {
            }
            n.oa = void 0;
          }
        } else
          _(n, 11);
      } else if ((t.K || n.g == t) && Tt(n), !W(i))
        for (r = n.Ja.g.parse(i), i = 0; i < r.length; i++) {
          let u = r[i];
          if (n.V = u[0], u = u[1], n.H == 2)
            if (u[0] == "c") {
              n.K = u[1], n.pa = u[2];
              const y = u[3];
              y != null && (n.ra = y, n.l.info("VER=" + n.ra));
              const P = u[4];
              P != null && (n.Ga = P, n.l.info("SVER=" + n.Ga));
              const pt = u[5];
              pt != null && typeof pt == "number" && 0 < pt && (e = 1.5 * pt, n.L = e, n.l.info("backChannelRequestTimeoutMs_=" + e)), e = n;
              const G = t.g;
              if (G) {
                const ct = G.g ? G.g.getResponseHeader("X-Client-Wire-Protocol") : null;
                if (ct) {
                  var s = e.i;
                  s.g || ct.indexOf("spdy") == -1 && ct.indexOf("quic") == -1 && ct.indexOf("h2") == -1 || (s.j = s.l, s.g = /* @__PURE__ */ new Set(), s.h && (vi(s, s.h), s.h = null));
                }
                if (e.F) {
                  const Xt = G.g ? G.g.getResponseHeader("X-HTTP-Session-Id") : null;
                  Xt && (e.Da = Xt, p(e.I, e.F, Xt));
                }
              }
              n.H = 3, n.h && n.h.Ba(), n.ca && (n.S = Date.now() - t.G, n.l.info("Handshake RTT: " + n.S + "ms")), e = n;
              var o = t;
              if (e.wa = Dn(e, e.J ? e.pa : null, e.Y), o.K) {
                dn(e.i, o);
                var f = o, a = e.L;
                a && f.setTimeout(a), f.C && (Ot(f), ut(f)), e.g = o;
              } else
                Mn(e);
              0 < n.j.length && Ht(n);
            } else
              u[0] != "stop" && u[0] != "close" || _(n, 7);
          else
            n.H == 3 && (u[0] == "stop" || u[0] == "close" ? u[0] == "stop" ? _(n, 7) : wi(n) : u[0] != "noop" && n.h && n.h.Aa(u), n.A = 0);
        }
    }
    it(4);
  } catch {
  }
}
function de(t) {
  if (t.Z && typeof t.Z == "function")
    return t.Z();
  if (typeof Map < "u" && t instanceof Map || typeof Set < "u" && t instanceof Set)
    return Array.from(t.values());
  if (typeof t == "string")
    return t.split("");
  if (jt(t)) {
    for (var i = [], n = t.length, e = 0; e < n; e++)
      i.push(t[e]);
    return i;
  }
  i = [], n = 0;
  for (e in t)
    i[n++] = t[e];
  return i;
}
function me(t) {
  if (t.ta && typeof t.ta == "function")
    return t.ta();
  if (!t.Z || typeof t.Z != "function") {
    if (typeof Map < "u" && t instanceof Map)
      return Array.from(t.keys());
    if (!(typeof Set < "u" && t instanceof Set)) {
      if (jt(t) || typeof t == "string") {
        var i = [];
        t = t.length;
        for (var n = 0; n < t; n++)
          i.push(n);
        return i;
      }
      i = [], n = 0;
      for (const e in t)
        i[n++] = e;
      return i;
    }
  }
}
function un(t, i) {
  if (t.forEach && typeof t.forEach == "function")
    t.forEach(i, void 0);
  else if (jt(t) || typeof t == "string")
    Array.prototype.forEach.call(t, i, void 0);
  else
    for (var n = me(t), e = de(t), r = e.length, s = 0; s < r; s++)
      i.call(void 0, e[s], n && n[s], t);
}
var fn = RegExp("^(?:([^:/?#.]+):)?(?://(?:([^\\\\/?#]*)@)?([^\\\\/?#]*?)(?::([0-9]+))?(?=[\\\\/?#]|$))?([^?#]+)?(?:\\?([^#]*))?(?:#([\\s\\S]*))?$");
function we(t, i) {
  if (t) {
    t = t.split("&");
    for (var n = 0; n < t.length; n++) {
      var e = t[n].indexOf("="), r = null;
      if (0 <= e) {
        var s = t[n].substring(0, e);
        r = t[n].substring(e + 1);
      } else
        s = t[n];
      i(s, r ? decodeURIComponent(r.replace(/\+/g, " ")) : "");
    }
  }
}
function H(t) {
  if (this.g = this.s = this.j = "", this.m = null, this.o = this.l = "", this.h = !1, t instanceof H) {
    this.h = t.h, mt(this, t.j), this.s = t.s, this.g = t.g, wt(this, t.m), this.l = t.l;
    var i = t.i, n = new nt();
    n.i = i.i, i.g && (n.g = new Map(i.g), n.h = i.h), Oi(this, n), this.o = t.o;
  } else
    t && (i = String(t).match(fn)) ? (this.h = !1, mt(this, i[1] || "", !0), this.s = z(i[2] || ""), this.g = z(i[3] || "", !0), wt(this, i[4]), this.l = z(i[5] || "", !0), Oi(this, i[6] || "", !0), this.o = z(i[7] || "")) : (this.h = !1, this.i = new nt(null, this.h));
}
H.prototype.toString = function() {
  var t = [], i = this.j;
  i && t.push(q(i, Pi, !0), ":");
  var n = this.g;
  return (n || i == "file") && (t.push("//"), (i = this.s) && t.push(q(i, Pi, !0), "@"), t.push(encodeURIComponent(String(n)).replace(/%25([0-9a-fA-F]{2})/g, "%$1")), n = this.m, n != null && t.push(":", String(n))), (n = this.l) && (this.g && n.charAt(0) != "/" && t.push("/"), t.push(q(n, n.charAt(0) == "/" ? Se : Ee, !0))), (n = this.i.toString()) && t.push("?", n), (n = this.o) && t.push("#", q(n, Ae)), t.join("");
};
function O(t) {
  return new H(t);
}
function mt(t, i, n) {
  t.j = n ? z(i, !0) : i, t.j && (t.j = t.j.replace(/:$/, ""));
}
function wt(t, i) {
  if (i) {
    if (i = Number(i), isNaN(i) || 0 > i)
      throw Error("Bad port number " + i);
    t.m = i;
  } else
    t.m = null;
}
function Oi(t, i, n) {
  i instanceof nt ? (t.i = i, Ce(t.i, t.h)) : (n || (i = q(i, je)), t.i = new nt(i, t.h));
}
function p(t, i, n) {
  t.i.set(i, n);
}
function Pt(t) {
  return p(t, "zx", Math.floor(2147483648 * Math.random()).toString(36) + Math.abs(Math.floor(2147483648 * Math.random()) ^ Date.now()).toString(36)), t;
}
function z(t, i) {
  return t ? i ? decodeURI(t.replace(/%25/g, "%2525")) : decodeURIComponent(t) : "";
}
function q(t, i, n) {
  return typeof t == "string" ? (t = encodeURI(t).replace(i, Te), n && (t = t.replace(/%25([0-9a-fA-F]{2})/g, "%$1")), t) : null;
}
function Te(t) {
  return t = t.charCodeAt(0), "%" + (t >> 4 & 15).toString(16) + (t & 15).toString(16);
}
var Pi = /[#\/\?@]/g, Ee = /[#\?:]/g, Se = /[#\?]/g, je = /[#\?@]/g, Ae = /#/g;
function nt(t, i) {
  this.h = this.g = null, this.i = t || null, this.j = !!i;
}
function D(t) {
  t.g || (t.g = /* @__PURE__ */ new Map(), t.h = 0, t.i && we(t.i, function(i, n) {
    t.add(decodeURIComponent(i.replace(/\+/g, " ")), n);
  }));
}
h = nt.prototype;
h.add = function(t, i) {
  D(this), this.i = null, t = U(this, t);
  var n = this.g.get(t);
  return n || this.g.set(t, n = []), n.push(i), this.h += 1, this;
};
function gn(t, i) {
  D(t), i = U(t, i), t.g.has(i) && (t.i = null, t.h -= t.g.get(i).length, t.g.delete(i));
}
function pn(t, i) {
  return D(t), i = U(t, i), t.g.has(i);
}
h.forEach = function(t, i) {
  D(this), this.g.forEach(function(n, e) {
    n.forEach(function(r) {
      t.call(i, r, e, this);
    }, this);
  }, this);
};
h.ta = function() {
  D(this);
  const t = Array.from(this.g.values()), i = Array.from(this.g.keys()), n = [];
  for (let e = 0; e < i.length; e++) {
    const r = t[e];
    for (let s = 0; s < r.length; s++)
      n.push(i[e]);
  }
  return n;
};
h.Z = function(t) {
  D(this);
  let i = [];
  if (typeof t == "string")
    pn(this, t) && (i = i.concat(this.g.get(U(this, t))));
  else {
    t = Array.from(this.g.values());
    for (let n = 0; n < t.length; n++)
      i = i.concat(t[n]);
  }
  return i;
};
h.set = function(t, i) {
  return D(this), this.i = null, t = U(this, t), pn(this, t) && (this.h -= this.g.get(t).length), this.g.set(t, [i]), this.h += 1, this;
};
h.get = function(t, i) {
  return t ? (t = this.Z(t), 0 < t.length ? String(t[0]) : i) : i;
};
function cn(t, i, n) {
  gn(t, i), 0 < n.length && (t.i = null, t.g.set(U(t, i), ti(n)), t.h += n.length);
}
h.toString = function() {
  if (this.i)
    return this.i;
  if (!this.g)
    return "";
  const t = [], i = Array.from(this.g.keys());
  for (var n = 0; n < i.length; n++) {
    var e = i[n];
    const s = encodeURIComponent(String(e)), o = this.Z(e);
    for (e = 0; e < o.length; e++) {
      var r = s;
      o[e] !== "" && (r += "=" + encodeURIComponent(String(o[e]))), t.push(r);
    }
  }
  return this.i = t.join("&");
};
function U(t, i) {
  return i = String(i), t.j && (i = i.toLowerCase()), i;
}
function Ce(t, i) {
  i && !t.j && (D(t), t.i = null, t.g.forEach(function(n, e) {
    var r = e.toLowerCase();
    e != r && (gn(this, e), cn(this, r, n));
  }, t)), t.j = i;
}
var xe = class {
  constructor(t, i) {
    this.g = t, this.map = i;
  }
};
function an(t) {
  this.l = t || Ie, l.PerformanceNavigationTiming ? (t = l.performance.getEntriesByType("navigation"), t = 0 < t.length && (t[0].nextHopProtocol == "hq" || t[0].nextHopProtocol == "h2")) : t = !!(l.g && l.g.Ka && l.g.Ka() && l.g.Ka().dc), this.j = t ? this.l : 1, this.g = null, 1 < this.j && (this.g = /* @__PURE__ */ new Set()), this.h = null, this.i = [];
}
var Ie = 10;
function yn(t) {
  return t.h ? !0 : t.g ? t.g.size >= t.j : !1;
}
function vn(t) {
  return t.h ? 1 : t.g ? t.g.size : 0;
}
function $t(t, i) {
  return t.h ? t.h == i : t.g ? t.g.has(i) : !1;
}
function vi(t, i) {
  t.g ? t.g.add(i) : t.h = i;
}
function dn(t, i) {
  t.h && t.h == i ? t.h = null : t.g && t.g.has(i) && t.g.delete(i);
}
an.prototype.cancel = function() {
  if (this.i = mn(this), this.h)
    this.h.cancel(), this.h = null;
  else if (this.g && this.g.size !== 0) {
    for (const t of this.g.values())
      t.cancel();
    this.g.clear();
  }
};
function mn(t) {
  if (t.h != null)
    return t.i.concat(t.h.F);
  if (t.g != null && t.g.size !== 0) {
    let i = t.i;
    for (const n of t.g.values())
      i = i.concat(n.F);
    return i;
  }
  return ti(t.i);
}
var Me = class {
  stringify(t) {
    return l.JSON.stringify(t, void 0);
  }
  parse(t) {
    return l.JSON.parse(t, void 0);
  }
};
function Re() {
  this.g = new Me();
}
function Oe(t, i, n) {
  const e = n || "";
  try {
    un(t, function(r, s) {
      let o = r;
      rt(r) && (o = hi(r)), i.push(e + s + "=" + encodeURIComponent(o));
    });
  } catch (r) {
    throw i.push(e + "type=" + encodeURIComponent("_badmap")), r;
  }
}
function Pe(t, i) {
  const n = new Mt();
  if (l.Image) {
    const e = new Image();
    e.onload = at(vt, n, e, "TestLoadImage: loaded", !0, i), e.onerror = at(vt, n, e, "TestLoadImage: error", !1, i), e.onabort = at(vt, n, e, "TestLoadImage: abort", !1, i), e.ontimeout = at(vt, n, e, "TestLoadImage: timeout", !1, i), l.setTimeout(function() {
      e.ontimeout && e.ontimeout();
    }, 1e4), e.src = t;
  } else
    i(!1);
}
function vt(t, i, n, e, r) {
  try {
    i.onload = null, i.onerror = null, i.onabort = null, i.ontimeout = null, r(e);
  } catch {
  }
}
function kt(t) {
  this.l = t.ec || null, this.j = t.ob || !1;
}
d(kt, ci);
kt.prototype.g = function() {
  return new Dt(this.l, this.j);
};
kt.prototype.i = function(t) {
  return function() {
    return t;
  };
}({});
function Dt(t, i) {
  v.call(this), this.F = t, this.u = i, this.m = void 0, this.readyState = di, this.status = 0, this.responseType = this.responseText = this.response = this.statusText = "", this.onreadystatechange = null, this.v = new Headers(), this.h = null, this.C = "GET", this.B = "", this.g = !1, this.A = this.j = this.l = null;
}
d(Dt, v);
var di = 0;
h = Dt.prototype;
h.open = function(t, i) {
  if (this.readyState != di)
    throw this.abort(), Error("Error reopening a connection");
  this.C = t, this.B = i, this.readyState = 1, et(this);
};
h.send = function(t) {
  if (this.readyState != 1)
    throw this.abort(), Error("need to call open() first. ");
  this.g = !0;
  const i = { headers: this.v, method: this.C, credentials: this.m, cache: void 0 };
  t && (i.body = t), (this.F || l).fetch(new Request(this.B, i)).then(this.$a.bind(this), this.ka.bind(this));
};
h.abort = function() {
  this.response = this.responseText = "", this.v = new Headers(), this.status = 0, this.j && this.j.cancel("Request was aborted.").catch(() => {
  }), 1 <= this.readyState && this.g && this.readyState != 4 && (this.g = !1, ft(this)), this.readyState = di;
};
h.$a = function(t) {
  if (this.g && (this.l = t, this.h || (this.status = this.l.status, this.statusText = this.l.statusText, this.h = t.headers, this.readyState = 2, et(this)), this.g && (this.readyState = 3, et(this), this.g)))
    if (this.responseType === "arraybuffer")
      t.arrayBuffer().then(this.Ya.bind(this), this.ka.bind(this));
    else if (typeof l.ReadableStream < "u" && "body" in t) {
      if (this.j = t.body.getReader(), this.u) {
        if (this.responseType)
          throw Error('responseType must be empty for "streamBinaryChunks" mode responses.');
        this.response = [];
      } else
        this.response = this.responseText = "", this.A = new TextDecoder();
      wn(this);
    } else
      t.text().then(this.Za.bind(this), this.ka.bind(this));
};
function wn(t) {
  t.j.read().then(t.Xa.bind(t)).catch(t.ka.bind(t));
}
h.Xa = function(t) {
  if (this.g) {
    if (this.u && t.value)
      this.response.push(t.value);
    else if (!this.u) {
      var i = t.value ? t.value : new Uint8Array(0);
      (i = this.A.decode(i, { stream: !t.done })) && (this.response = this.responseText += i);
    }
    t.done ? ft(this) : et(this), this.readyState == 3 && wn(this);
  }
};
h.Za = function(t) {
  this.g && (this.response = this.responseText = t, ft(this));
};
h.Ya = function(t) {
  this.g && (this.response = t, ft(this));
};
h.ka = function() {
  this.g && ft(this);
};
function ft(t) {
  t.readyState = 4, t.l = null, t.j = null, t.A = null, et(t);
}
h.setRequestHeader = function(t, i) {
  this.v.append(t, i);
};
h.getResponseHeader = function(t) {
  return this.h && this.h.get(t.toLowerCase()) || "";
};
h.getAllResponseHeaders = function() {
  if (!this.h)
    return "";
  const t = [], i = this.h.entries();
  for (var n = i.next(); !n.done; )
    n = n.value, t.push(n[0] + ": " + n[1]), n = i.next();
  return t.join(`\r
`);
};
function et(t) {
  t.onreadystatechange && t.onreadystatechange.call(t);
}
Object.defineProperty(Dt.prototype, "withCredentials", { get: function() {
  return this.m === "include";
}, set: function(t) {
  this.m = t ? "include" : "same-origin";
} });
var ke = l.JSON.parse;
function c(t) {
  v.call(this), this.headers = /* @__PURE__ */ new Map(), this.u = t || null, this.h = !1, this.C = this.g = null, this.I = "", this.m = 0, this.j = "", this.l = this.G = this.v = this.F = !1, this.B = 0, this.A = null, this.K = Tn, this.L = this.M = !1;
}
d(c, v);
var Tn = "", De = /^https?$/i, Le = ["POST", "PUT"];
h = c.prototype;
h.Oa = function(t) {
  this.M = t;
};
h.ha = function(t, i, n, e) {
  if (this.g)
    throw Error("[goog.net.XhrIo] Object is active with another request=" + this.I + "; newUri=" + t);
  i = i ? i.toUpperCase() : "GET", this.I = t, this.j = "", this.m = 0, this.F = !1, this.h = !0, this.g = this.u ? this.u.g() : zt.g(), this.C = this.u ? Ri(this.u) : Ri(zt), this.g.onreadystatechange = T(this.La, this);
  try {
    this.G = !0, this.g.open(i, String(t), !0), this.G = !1;
  } catch (s) {
    ki(this, s);
    return;
  }
  if (t = n || "", n = new Map(this.headers), e)
    if (Object.getPrototypeOf(e) === Object.prototype)
      for (var r in e)
        n.set(r, e[r]);
    else if (typeof e.keys == "function" && typeof e.get == "function")
      for (const s of e.keys())
        n.set(s, e.get(s));
    else
      throw Error("Unknown input type for opt_headers: " + String(e));
  e = Array.from(n.keys()).find((s) => s.toLowerCase() == "content-type"), r = l.FormData && t instanceof l.FormData, !(0 <= Xi(Le, i)) || e || r || n.set("Content-Type", "application/x-www-form-urlencoded;charset=utf-8");
  for (const [s, o] of n)
    this.g.setRequestHeader(s, o);
  this.K && (this.g.responseType = this.K), "withCredentials" in this.g && this.g.withCredentials !== this.M && (this.g.withCredentials = this.M);
  try {
    jn(this), 0 < this.B && ((this.L = _e(this.g)) ? (this.g.timeout = this.B, this.g.ontimeout = T(this.ua, this)) : this.A = fi(this.ua, this.B, this)), this.v = !0, this.g.send(t), this.v = !1;
  } catch (s) {
    ki(this, s);
  }
};
function _e(t) {
  return $ && typeof t.timeout == "number" && t.ontimeout !== void 0;
}
h.ua = function() {
  typeof bt < "u" && this.g && (this.j = "Timed out after " + this.B + "ms, aborting", this.m = 8, w(this, "timeout"), this.abort(8));
};
function ki(t, i) {
  t.h = !1, t.g && (t.l = !0, t.g.abort(), t.l = !1), t.j = i, t.m = 5, En(t), Lt(t);
}
function En(t) {
  t.F || (t.F = !0, w(t, "complete"), w(t, "error"));
}
h.abort = function(t) {
  this.g && this.h && (this.h = !1, this.l = !0, this.g.abort(), this.l = !1, this.m = t || 7, w(this, "complete"), w(this, "abort"), Lt(this));
};
h.N = function() {
  this.g && (this.h && (this.h = !1, this.l = !0, this.g.abort(), this.l = !1), Lt(this, !0)), c.$.N.call(this);
};
h.La = function() {
  this.s || (this.G || this.v || this.l ? Sn(this) : this.kb());
};
h.kb = function() {
  Sn(this);
};
function Sn(t) {
  if (t.h && typeof bt < "u" && (!t.C[1] || x(t) != 4 || t.da() != 2)) {
    if (t.v && x(t) == 4)
      fi(t.La, 0, t);
    else if (w(t, "readystatechange"), x(t) == 4) {
      t.h = !1;
      try {
        const o = t.da();
        t:
          switch (o) {
            case 200:
            case 201:
            case 202:
            case 204:
            case 206:
            case 304:
            case 1223:
              var i = !0;
              break t;
            default:
              i = !1;
          }
        var n;
        if (!(n = i)) {
          var e;
          if (e = o === 0) {
            var r = String(t.I).match(fn)[1] || null;
            !r && l.self && l.self.location && (r = l.self.location.protocol.slice(0, -1)), e = !De.test(r ? r.toLowerCase() : "");
          }
          n = e;
        }
        if (n)
          w(t, "complete"), w(t, "success");
        else {
          t.m = 6;
          try {
            var s = 2 < x(t) ? t.g.statusText : "";
          } catch {
            s = "";
          }
          t.j = s + " [" + t.da() + "]", En(t);
        }
      } finally {
        Lt(t);
      }
    }
  }
}
function Lt(t, i) {
  if (t.g) {
    jn(t);
    const n = t.g, e = t.C[0] ? () => {
    } : null;
    t.g = null, t.C = null, i || w(t, "ready");
    try {
      n.onreadystatechange = e;
    } catch {
    }
  }
}
function jn(t) {
  t.g && t.L && (t.g.ontimeout = null), t.A && (l.clearTimeout(t.A), t.A = null);
}
h.isActive = function() {
  return !!this.g;
};
function x(t) {
  return t.g ? t.g.readyState : 0;
}
h.da = function() {
  try {
    return 2 < x(this) ? this.g.status : -1;
  } catch {
    return -1;
  }
};
h.ja = function() {
  try {
    return this.g ? this.g.responseText : "";
  } catch {
    return "";
  }
};
h.Wa = function(t) {
  if (this.g) {
    var i = this.g.responseText;
    return t && i.indexOf(t) == 0 && (i = i.substring(t.length)), ke(i);
  }
};
function Di(t) {
  try {
    if (!t.g)
      return null;
    if ("response" in t.g)
      return t.g.response;
    switch (t.K) {
      case Tn:
      case "text":
        return t.g.responseText;
      case "arraybuffer":
        if ("mozResponseArrayBuffer" in t.g)
          return t.g.mozResponseArrayBuffer;
    }
    return null;
  } catch {
    return null;
  }
}
function He(t) {
  const i = {};
  t = (t.g && 2 <= x(t) && t.g.getAllResponseHeaders() || "").split(`\r
`);
  for (let e = 0; e < t.length; e++) {
    if (W(t[e]))
      continue;
    var n = se(t[e]);
    const r = n[0];
    if (n = n[1], typeof n != "string")
      continue;
    n = n.trim();
    const s = i[r] || [];
    i[r] = s, s.push(n);
  }
  Qn(i, function(e) {
    return e.join(", ");
  });
}
h.Ia = function() {
  return this.m;
};
h.Sa = function() {
  return typeof this.j == "string" ? this.j : String(this.j);
};
function An(t) {
  let i = "";
  return ni(t, function(n, e) {
    i += e, i += ":", i += n, i += `\r
`;
  }), i;
}
function mi(t, i, n) {
  t: {
    for (e in n) {
      var e = !1;
      break t;
    }
    e = !0;
  }
  e || (n = An(n), typeof t == "string" ? n != null && encodeURIComponent(String(n)) : p(t, i, n));
}
function J(t, i, n) {
  return n && n.internalChannelParams && n.internalChannelParams[t] || i;
}
function Cn(t) {
  this.Ga = 0, this.j = [], this.l = new Mt(), this.pa = this.wa = this.I = this.Y = this.g = this.Da = this.F = this.na = this.o = this.U = this.s = null, this.fb = this.W = 0, this.cb = J("failFast", !1, t), this.G = this.v = this.u = this.m = this.h = null, this.aa = !0, this.Fa = this.V = -1, this.ba = this.A = this.C = 0, this.ab = J("baseRetryDelayMs", 5e3, t), this.hb = J("retryDelaySeedMs", 1e4, t), this.eb = J("forwardChannelMaxRetries", 2, t), this.xa = J("forwardChannelRequestTimeoutMs", 2e4, t), this.va = t && t.xmlHttpFactory || void 0, this.Ha = t && t.useFetchStreams || !1, this.L = void 0, this.J = t && t.supportsCrossDomainXhr || !1, this.K = "", this.i = new an(t && t.concurrentRequestLimit), this.Ja = new Re(), this.P = t && t.fastHandshake || !1, this.O = t && t.encodeInitMessageHeaders || !1, this.P && this.O && (this.O = !1), this.bb = t && t.bc || !1, t && t.Ea && this.l.Ea(), t && t.forceLongPolling && (this.aa = !1), this.ca = !this.P && this.aa && t && t.detectBufferingProxy || !1, this.qa = void 0, t && t.longPollingTimeout && 0 < t.longPollingTimeout && (this.qa = t.longPollingTimeout), this.oa = void 0, this.S = 0, this.M = !1, this.ma = this.B = null;
}
h = Cn.prototype;
h.ra = 8;
h.H = 1;
function wi(t) {
  if (xn(t), t.H == 3) {
    var i = t.W++, n = O(t.I);
    if (p(n, "SID", t.K), p(n, "RID", i), p(n, "TYPE", "terminate"), gt(t, n), i = new lt(t, t.l, i), i.L = 2, i.A = Pt(O(n)), n = !1, l.navigator && l.navigator.sendBeacon)
      try {
        n = l.navigator.sendBeacon(i.A.toString(), "");
      } catch {
      }
    !n && l.Image && (new Image().src = i.A, n = !0), n || (i.g = Ln(i.l, null), i.g.ha(i.A)), i.G = Date.now(), ut(i);
  }
  kn(t);
}
function _t(t) {
  t.g && (Ei(t), t.g.cancel(), t.g = null);
}
function xn(t) {
  _t(t), t.u && (l.clearTimeout(t.u), t.u = null), Tt(t), t.i.cancel(), t.m && (typeof t.m == "number" && l.clearTimeout(t.m), t.m = null);
}
function Ht(t) {
  if (!yn(t.i) && !t.m) {
    t.m = !0;
    var i = t.Na;
    Q || Wi(), b || (Q(), b = !0), li.add(i, t), t.C = 0;
  }
}
function Xe(t, i) {
  return vn(t.i) >= t.i.j - (t.m ? 1 : 0) ? !1 : t.m ? (t.j = i.F.concat(t.j), !0) : t.H == 1 || t.H == 2 || t.C >= (t.cb ? 0 : t.eb) ? !1 : (t.m = ot(T(t.Na, t, i), Pn(t, t.C)), t.C++, !0);
}
h.Na = function(t) {
  if (this.m)
    if (this.m = null, this.H == 1) {
      if (!t) {
        this.W = Math.floor(1e5 * Math.random()), t = this.W++;
        const r = new lt(this, this.l, t);
        let s = this.s;
        if (this.U && (s ? (s = Ui(s), Ki(s, this.U)) : s = this.U), this.o !== null || this.O || (r.I = s, s = null), this.P)
          t: {
            for (var i = 0, n = 0; n < this.j.length; n++) {
              i: {
                var e = this.j[n];
                if ("__data__" in e.map && (e = e.map.__data__, typeof e == "string")) {
                  e = e.length;
                  break i;
                }
                e = void 0;
              }
              if (e === void 0)
                break;
              if (i += e, 4096 < i) {
                i = n;
                break t;
              }
              if (i === 4096 || n === this.j.length - 1) {
                i = n + 1;
                break t;
              }
            }
            i = 1e3;
          }
        else
          i = 1e3;
        i = In(this, r, i), n = O(this.I), p(n, "RID", t), p(n, "CVER", 22), this.F && p(n, "X-HTTP-Session-Id", this.F), gt(this, n), s && (this.O ? i = "headers=" + encodeURIComponent(String(An(s))) + "&" + i : this.o && mi(n, this.o, s)), vi(this.i, r), this.bb && p(n, "TYPE", "init"), this.P ? (p(n, "$req", i), p(n, "SID", "null"), r.aa = !0, Vt(r, n, null)) : Vt(r, n, i), this.H = 2;
      }
    } else
      this.H == 3 && (t ? Li(this, t) : this.j.length == 0 || yn(this.i) || Li(this));
};
function Li(t, i) {
  var n;
  i ? n = i.m : n = t.W++;
  const e = O(t.I);
  p(e, "SID", t.K), p(e, "RID", n), p(e, "AID", t.V), gt(t, e), t.o && t.s && mi(e, t.o, t.s), n = new lt(t, t.l, n, t.C + 1), t.o === null && (n.I = t.s), i && (t.j = i.F.concat(t.j)), i = In(t, n, 1e3), n.setTimeout(Math.round(0.5 * t.xa) + Math.round(0.5 * t.xa * Math.random())), vi(t.i, n), Vt(n, e, i);
}
function gt(t, i) {
  t.na && ni(t.na, function(n, e) {
    p(i, e, n);
  }), t.h && un({}, function(n, e) {
    p(i, e, n);
  });
}
function In(t, i, n) {
  n = Math.min(t.j.length, n);
  var e = t.h ? T(t.h.Va, t.h, t) : null;
  t: {
    var r = t.j;
    let s = -1;
    for (; ; ) {
      const o = ["count=" + n];
      s == -1 ? 0 < n ? (s = r[0].g, o.push("ofs=" + s)) : s = 0 : o.push("ofs=" + s);
      let f = !0;
      for (let a = 0; a < n; a++) {
        let u = r[a].g;
        const y = r[a].map;
        if (u -= s, 0 > u)
          s = Math.max(0, r[a].g - 100), f = !1;
        else
          try {
            Oe(y, o, "req" + u + "_");
          } catch {
            e && e(y);
          }
      }
      if (f) {
        e = o.join("&");
        break t;
      }
    }
  }
  return t = t.j.splice(0, n), i.F = t, e;
}
function Mn(t) {
  if (!t.g && !t.u) {
    t.ba = 1;
    var i = t.Ma;
    Q || Wi(), b || (Q(), b = !0), li.add(i, t), t.A = 0;
  }
}
function Ti(t) {
  return t.g || t.u || 3 <= t.A ? !1 : (t.ba++, t.u = ot(T(t.Ma, t), Pn(t, t.A)), t.A++, !0);
}
h.Ma = function() {
  if (this.u = null, Rn(this), this.ca && !(this.M || this.g == null || 0 >= this.S)) {
    var t = 2 * this.S;
    this.l.info("BP detection timer enabled: " + t), this.B = ot(T(this.jb, this), t);
  }
};
h.jb = function() {
  this.B && (this.B = null, this.l.info("BP detection timeout reached."), this.l.info("Buffering proxy detected and switch to long-polling!"), this.G = !1, this.M = !0, S(10), _t(this), Rn(this));
};
function Ei(t) {
  t.B != null && (l.clearTimeout(t.B), t.B = null);
}
function Rn(t) {
  t.g = new lt(t, t.l, "rpc", t.ba), t.o === null && (t.g.I = t.s), t.g.O = 0;
  var i = O(t.wa);
  p(i, "RID", "rpc"), p(i, "SID", t.K), p(i, "AID", t.V), p(i, "CI", t.G ? "0" : "1"), !t.G && t.qa && p(i, "TO", t.qa), p(i, "TYPE", "xmlhttp"), gt(t, i), t.o && t.s && mi(i, t.o, t.s), t.L && t.g.setTimeout(t.L);
  var n = t.g;
  t = t.pa, n.L = 1, n.A = Pt(O(i)), n.u = null, n.S = !0, sn(n, t);
}
h.ib = function() {
  this.v != null && (this.v = null, _t(this), Ti(this), S(19));
};
function Tt(t) {
  t.v != null && (l.clearTimeout(t.v), t.v = null);
}
function On(t, i) {
  var n = null;
  if (t.g == i) {
    Tt(t), Ei(t), t.g = null;
    var e = 2;
  } else if ($t(t.i, i))
    n = i.F, dn(t.i, i), e = 1;
  else
    return;
  if (t.H != 0) {
    if (i.i)
      if (e == 1) {
        n = i.u ? i.u.length : 0, i = Date.now() - i.G;
        var r = t.C;
        e = gi(), w(e, new nn(e, n)), Ht(t);
      } else
        Mn(t);
    else if (r = i.s, r == 3 || r == 0 && 0 < i.ca || !(e == 1 && Xe(t, i) || e == 2 && Ti(t)))
      switch (n && 0 < n.length && (i = t.i, i.i = i.i.concat(n)), r) {
        case 1:
          _(t, 5);
          break;
        case 4:
          _(t, 10);
          break;
        case 3:
          _(t, 6);
          break;
        default:
          _(t, 2);
      }
  }
}
function Pn(t, i) {
  let n = t.ab + Math.floor(Math.random() * t.hb);
  return t.isActive() || (n *= 2), n * i;
}
function _(t, i) {
  if (t.l.info("Error code " + i), i == 2) {
    var n = null;
    t.h && (n = null);
    var e = T(t.pb, t);
    n || (n = new H("//www.google.com/images/cleardot.gif"), l.location && l.location.protocol == "http" || mt(n, "https"), Pt(n)), Pe(n.toString(), e);
  } else
    S(2);
  t.H = 0, t.h && t.h.za(i), kn(t), xn(t);
}
h.pb = function(t) {
  t ? (this.l.info("Successfully pinged google.com"), S(2)) : (this.l.info("Failed to ping google.com"), S(1));
};
function kn(t) {
  if (t.H = 0, t.ma = [], t.h) {
    const i = mn(t.i);
    (i.length != 0 || t.j.length != 0) && (ji(t.ma, i), ji(t.ma, t.j), t.i.i.length = 0, ti(t.j), t.j.length = 0), t.h.ya();
  }
}
function Dn(t, i, n) {
  var e = n instanceof H ? O(n) : new H(n);
  if (e.g != "")
    i && (e.g = i + "." + e.g), wt(e, e.m);
  else {
    var r = l.location;
    e = r.protocol, i = i ? i + "." + r.hostname : r.hostname, r = +r.port;
    var s = new H(null);
    e && mt(s, e), i && (s.g = i), r && wt(s, r), n && (s.l = n), e = s;
  }
  return n = t.F, i = t.Da, n && i && p(e, n, i), p(e, "VER", t.ra), gt(t, e), e;
}
function Ln(t, i, n) {
  if (i && !t.J)
    throw Error("Can't create secondary domain capable XhrIo object.");
  return i = t.Ha && !t.va ? new c(new kt({ ob: n })) : new c(t.va), i.Oa(t.J), i;
}
h.isActive = function() {
  return !!this.h && this.h.isActive(this);
};
function _n() {
}
h = _n.prototype;
h.Ba = function() {
};
h.Aa = function() {
};
h.za = function() {
};
h.ya = function() {
};
h.isActive = function() {
  return !0;
};
h.Va = function() {
};
function A(t, i) {
  v.call(this), this.g = new Cn(i), this.l = t, this.h = i && i.messageUrlParams || null, t = i && i.messageHeaders || null, i && i.clientProtocolHeaderRequired && (t ? t["X-Client-Protocol"] = "webchannel" : t = { "X-Client-Protocol": "webchannel" }), this.g.s = t, t = i && i.initMessageHeaders || null, i && i.messageContentType && (t ? t["X-WebChannel-Content-Type"] = i.messageContentType : t = { "X-WebChannel-Content-Type": i.messageContentType }), i && i.Ca && (t ? t["X-WebChannel-Client-Profile"] = i.Ca : t = { "X-WebChannel-Client-Profile": i.Ca }), this.g.U = t, (t = i && i.cc) && !W(t) && (this.g.o = t), this.A = i && i.supportsCrossDomainXhr || !1, this.v = i && i.sendRawJson || !1, (i = i && i.httpSessionIdParam) && !W(i) && (this.g.F = i, t = this.h, t !== null && i in t && (t = this.h, i in t && delete t[i])), this.j = new K(this);
}
d(A, v);
A.prototype.m = function() {
  this.g.h = this.j, this.A && (this.g.J = !0);
  var t = this.g, i = this.l, n = this.h || void 0;
  S(0), t.Y = i, t.na = n || {}, t.G = t.aa, t.I = Dn(t, null, t.Y), Ht(t);
};
A.prototype.close = function() {
  wi(this.g);
};
A.prototype.u = function(t) {
  var i = this.g;
  if (typeof t == "string") {
    var n = {};
    n.__data__ = t, t = n;
  } else
    this.v && (n = {}, n.__data__ = hi(t), t = n);
  i.j.push(new xe(i.fb++, t)), i.H == 3 && Ht(i);
};
A.prototype.N = function() {
  this.g.h = null, delete this.j, wi(this.g), delete this.g, A.$.N.call(this);
};
function Hn(t) {
  ai.call(this), t.__headers__ && (this.headers = t.__headers__, this.statusCode = t.__status__, delete t.__headers__, delete t.__status__);
  var i = t.__sm__;
  if (i) {
    t: {
      for (const n in i) {
        t = n;
        break t;
      }
      t = void 0;
    }
    (this.i = t) && (t = this.i, i = i !== null && t in i ? i[t] : void 0), this.data = i;
  } else
    this.data = t;
}
d(Hn, ai);
function Xn() {
  yi.call(this), this.status = 1;
}
d(Xn, yi);
function K(t) {
  this.g = t;
}
d(K, _n);
K.prototype.Ba = function() {
  w(this.g, "a");
};
K.prototype.Aa = function(t) {
  w(this.g, new Hn(t));
};
K.prototype.za = function(t) {
  w(this.g, new Xn());
};
K.prototype.ya = function() {
  w(this.g, "b");
};
function Ne() {
  this.blockSize = -1;
}
function M() {
  this.blockSize = -1, this.blockSize = 64, this.g = Array(4), this.m = Array(this.blockSize), this.i = this.h = 0, this.reset();
}
d(M, Ne);
M.prototype.reset = function() {
  this.g[0] = 1732584193, this.g[1] = 4023233417, this.g[2] = 2562383102, this.g[3] = 271733878, this.i = this.h = 0;
};
function Kt(t, i, n) {
  n || (n = 0);
  var e = Array(16);
  if (typeof i == "string")
    for (var r = 0; 16 > r; ++r)
      e[r] = i.charCodeAt(n++) | i.charCodeAt(n++) << 8 | i.charCodeAt(n++) << 16 | i.charCodeAt(n++) << 24;
  else
    for (r = 0; 16 > r; ++r)
      e[r] = i[n++] | i[n++] << 8 | i[n++] << 16 | i[n++] << 24;
  i = t.g[0], n = t.g[1], r = t.g[2];
  var s = t.g[3], o = i + (s ^ n & (r ^ s)) + e[0] + 3614090360 & 4294967295;
  i = n + (o << 7 & 4294967295 | o >>> 25), o = s + (r ^ i & (n ^ r)) + e[1] + 3905402710 & 4294967295, s = i + (o << 12 & 4294967295 | o >>> 20), o = r + (n ^ s & (i ^ n)) + e[2] + 606105819 & 4294967295, r = s + (o << 17 & 4294967295 | o >>> 15), o = n + (i ^ r & (s ^ i)) + e[3] + 3250441966 & 4294967295, n = r + (o << 22 & 4294967295 | o >>> 10), o = i + (s ^ n & (r ^ s)) + e[4] + 4118548399 & 4294967295, i = n + (o << 7 & 4294967295 | o >>> 25), o = s + (r ^ i & (n ^ r)) + e[5] + 1200080426 & 4294967295, s = i + (o << 12 & 4294967295 | o >>> 20), o = r + (n ^ s & (i ^ n)) + e[6] + 2821735955 & 4294967295, r = s + (o << 17 & 4294967295 | o >>> 15), o = n + (i ^ r & (s ^ i)) + e[7] + 4249261313 & 4294967295, n = r + (o << 22 & 4294967295 | o >>> 10), o = i + (s ^ n & (r ^ s)) + e[8] + 1770035416 & 4294967295, i = n + (o << 7 & 4294967295 | o >>> 25), o = s + (r ^ i & (n ^ r)) + e[9] + 2336552879 & 4294967295, s = i + (o << 12 & 4294967295 | o >>> 20), o = r + (n ^ s & (i ^ n)) + e[10] + 4294925233 & 4294967295, r = s + (o << 17 & 4294967295 | o >>> 15), o = n + (i ^ r & (s ^ i)) + e[11] + 2304563134 & 4294967295, n = r + (o << 22 & 4294967295 | o >>> 10), o = i + (s ^ n & (r ^ s)) + e[12] + 1804603682 & 4294967295, i = n + (o << 7 & 4294967295 | o >>> 25), o = s + (r ^ i & (n ^ r)) + e[13] + 4254626195 & 4294967295, s = i + (o << 12 & 4294967295 | o >>> 20), o = r + (n ^ s & (i ^ n)) + e[14] + 2792965006 & 4294967295, r = s + (o << 17 & 4294967295 | o >>> 15), o = n + (i ^ r & (s ^ i)) + e[15] + 1236535329 & 4294967295, n = r + (o << 22 & 4294967295 | o >>> 10), o = i + (r ^ s & (n ^ r)) + e[1] + 4129170786 & 4294967295, i = n + (o << 5 & 4294967295 | o >>> 27), o = s + (n ^ r & (i ^ n)) + e[6] + 3225465664 & 4294967295, s = i + (o << 9 & 4294967295 | o >>> 23), o = r + (i ^ n & (s ^ i)) + e[11] + 643717713 & 4294967295, r = s + (o << 14 & 4294967295 | o >>> 18), o = n + (s ^ i & (r ^ s)) + e[0] + 3921069994 & 4294967295, n = r + (o << 20 & 4294967295 | o >>> 12), o = i + (r ^ s & (n ^ r)) + e[5] + 3593408605 & 4294967295, i = n + (o << 5 & 4294967295 | o >>> 27), o = s + (n ^ r & (i ^ n)) + e[10] + 38016083 & 4294967295, s = i + (o << 9 & 4294967295 | o >>> 23), o = r + (i ^ n & (s ^ i)) + e[15] + 3634488961 & 4294967295, r = s + (o << 14 & 4294967295 | o >>> 18), o = n + (s ^ i & (r ^ s)) + e[4] + 3889429448 & 4294967295, n = r + (o << 20 & 4294967295 | o >>> 12), o = i + (r ^ s & (n ^ r)) + e[9] + 568446438 & 4294967295, i = n + (o << 5 & 4294967295 | o >>> 27), o = s + (n ^ r & (i ^ n)) + e[14] + 3275163606 & 4294967295, s = i + (o << 9 & 4294967295 | o >>> 23), o = r + (i ^ n & (s ^ i)) + e[3] + 4107603335 & 4294967295, r = s + (o << 14 & 4294967295 | o >>> 18), o = n + (s ^ i & (r ^ s)) + e[8] + 1163531501 & 4294967295, n = r + (o << 20 & 4294967295 | o >>> 12), o = i + (r ^ s & (n ^ r)) + e[13] + 2850285829 & 4294967295, i = n + (o << 5 & 4294967295 | o >>> 27), o = s + (n ^ r & (i ^ n)) + e[2] + 4243563512 & 4294967295, s = i + (o << 9 & 4294967295 | o >>> 23), o = r + (i ^ n & (s ^ i)) + e[7] + 1735328473 & 4294967295, r = s + (o << 14 & 4294967295 | o >>> 18), o = n + (s ^ i & (r ^ s)) + e[12] + 2368359562 & 4294967295, n = r + (o << 20 & 4294967295 | o >>> 12), o = i + (n ^ r ^ s) + e[5] + 4294588738 & 4294967295, i = n + (o << 4 & 4294967295 | o >>> 28), o = s + (i ^ n ^ r) + e[8] + 2272392833 & 4294967295, s = i + (o << 11 & 4294967295 | o >>> 21), o = r + (s ^ i ^ n) + e[11] + 1839030562 & 4294967295, r = s + (o << 16 & 4294967295 | o >>> 16), o = n + (r ^ s ^ i) + e[14] + 4259657740 & 4294967295, n = r + (o << 23 & 4294967295 | o >>> 9), o = i + (n ^ r ^ s) + e[1] + 2763975236 & 4294967295, i = n + (o << 4 & 4294967295 | o >>> 28), o = s + (i ^ n ^ r) + e[4] + 1272893353 & 4294967295, s = i + (o << 11 & 4294967295 | o >>> 21), o = r + (s ^ i ^ n) + e[7] + 4139469664 & 4294967295, r = s + (o << 16 & 4294967295 | o >>> 16), o = n + (r ^ s ^ i) + e[10] + 3200236656 & 4294967295, n = r + (o << 23 & 4294967295 | o >>> 9), o = i + (n ^ r ^ s) + e[13] + 681279174 & 4294967295, i = n + (o << 4 & 4294967295 | o >>> 28), o = s + (i ^ n ^ r) + e[0] + 3936430074 & 4294967295, s = i + (o << 11 & 4294967295 | o >>> 21), o = r + (s ^ i ^ n) + e[3] + 3572445317 & 4294967295, r = s + (o << 16 & 4294967295 | o >>> 16), o = n + (r ^ s ^ i) + e[6] + 76029189 & 4294967295, n = r + (o << 23 & 4294967295 | o >>> 9), o = i + (n ^ r ^ s) + e[9] + 3654602809 & 4294967295, i = n + (o << 4 & 4294967295 | o >>> 28), o = s + (i ^ n ^ r) + e[12] + 3873151461 & 4294967295, s = i + (o << 11 & 4294967295 | o >>> 21), o = r + (s ^ i ^ n) + e[15] + 530742520 & 4294967295, r = s + (o << 16 & 4294967295 | o >>> 16), o = n + (r ^ s ^ i) + e[2] + 3299628645 & 4294967295, n = r + (o << 23 & 4294967295 | o >>> 9), o = i + (r ^ (n | ~s)) + e[0] + 4096336452 & 4294967295, i = n + (o << 6 & 4294967295 | o >>> 26), o = s + (n ^ (i | ~r)) + e[7] + 1126891415 & 4294967295, s = i + (o << 10 & 4294967295 | o >>> 22), o = r + (i ^ (s | ~n)) + e[14] + 2878612391 & 4294967295, r = s + (o << 15 & 4294967295 | o >>> 17), o = n + (s ^ (r | ~i)) + e[5] + 4237533241 & 4294967295, n = r + (o << 21 & 4294967295 | o >>> 11), o = i + (r ^ (n | ~s)) + e[12] + 1700485571 & 4294967295, i = n + (o << 6 & 4294967295 | o >>> 26), o = s + (n ^ (i | ~r)) + e[3] + 2399980690 & 4294967295, s = i + (o << 10 & 4294967295 | o >>> 22), o = r + (i ^ (s | ~n)) + e[10] + 4293915773 & 4294967295, r = s + (o << 15 & 4294967295 | o >>> 17), o = n + (s ^ (r | ~i)) + e[1] + 2240044497 & 4294967295, n = r + (o << 21 & 4294967295 | o >>> 11), o = i + (r ^ (n | ~s)) + e[8] + 1873313359 & 4294967295, i = n + (o << 6 & 4294967295 | o >>> 26), o = s + (n ^ (i | ~r)) + e[15] + 4264355552 & 4294967295, s = i + (o << 10 & 4294967295 | o >>> 22), o = r + (i ^ (s | ~n)) + e[6] + 2734768916 & 4294967295, r = s + (o << 15 & 4294967295 | o >>> 17), o = n + (s ^ (r | ~i)) + e[13] + 1309151649 & 4294967295, n = r + (o << 21 & 4294967295 | o >>> 11), o = i + (r ^ (n | ~s)) + e[4] + 4149444226 & 4294967295, i = n + (o << 6 & 4294967295 | o >>> 26), o = s + (n ^ (i | ~r)) + e[11] + 3174756917 & 4294967295, s = i + (o << 10 & 4294967295 | o >>> 22), o = r + (i ^ (s | ~n)) + e[2] + 718787259 & 4294967295, r = s + (o << 15 & 4294967295 | o >>> 17), o = n + (s ^ (r | ~i)) + e[9] + 3951481745 & 4294967295, t.g[0] = t.g[0] + i & 4294967295, t.g[1] = t.g[1] + (r + (o << 21 & 4294967295 | o >>> 11)) & 4294967295, t.g[2] = t.g[2] + r & 4294967295, t.g[3] = t.g[3] + s & 4294967295;
}
M.prototype.j = function(t, i) {
  i === void 0 && (i = t.length);
  for (var n = i - this.blockSize, e = this.m, r = this.h, s = 0; s < i; ) {
    if (r == 0)
      for (; s <= n; )
        Kt(this, t, s), s += this.blockSize;
    if (typeof t == "string") {
      for (; s < i; )
        if (e[r++] = t.charCodeAt(s++), r == this.blockSize) {
          Kt(this, e), r = 0;
          break;
        }
    } else
      for (; s < i; )
        if (e[r++] = t[s++], r == this.blockSize) {
          Kt(this, e), r = 0;
          break;
        }
  }
  this.h = r, this.i += i;
};
M.prototype.l = function() {
  var t = Array((56 > this.h ? this.blockSize : 2 * this.blockSize) - this.h);
  t[0] = 128;
  for (var i = 1; i < t.length - 8; ++i)
    t[i] = 0;
  var n = 8 * this.i;
  for (i = t.length - 8; i < t.length; ++i)
    t[i] = n & 255, n /= 256;
  for (this.j(t), t = Array(16), i = n = 0; 4 > i; ++i)
    for (var e = 0; 32 > e; e += 8)
      t[n++] = this.g[i] >>> e & 255;
  return t;
};
function g(t, i) {
  this.h = i;
  for (var n = [], e = !0, r = t.length - 1; 0 <= r; r--) {
    var s = t[r] | 0;
    e && s == i || (n[r] = s, e = !1);
  }
  this.g = n;
}
var Be = {};
function Si(t) {
  return -128 <= t && 128 > t ? zn(t, function(i) {
    return new g([i | 0], 0 > i ? -1 : 0);
  }) : new g([t | 0], 0 > t ? -1 : 0);
}
function I(t) {
  if (isNaN(t) || !isFinite(t))
    return B;
  if (0 > t)
    return m(I(-t));
  for (var i = [], n = 1, e = 0; t >= n; e++)
    i[e] = t / n | 0, n *= Zt;
  return new g(i, 0);
}
function Nn(t, i) {
  if (t.length == 0)
    throw Error("number format error: empty string");
  if (i = i || 10, 2 > i || 36 < i)
    throw Error("radix out of range: " + i);
  if (t.charAt(0) == "-")
    return m(Nn(t.substring(1), i));
  if (0 <= t.indexOf("-"))
    throw Error('number format error: interior "-" character');
  for (var n = I(Math.pow(i, 8)), e = B, r = 0; r < t.length; r += 8) {
    var s = Math.min(8, t.length - r), o = parseInt(t.substring(r, r + s), i);
    8 > s ? (s = I(Math.pow(i, s)), e = e.R(s).add(I(o))) : (e = e.R(n), e = e.add(I(o)));
  }
  return e;
}
var Zt = 4294967296, B = Si(0), Qt = Si(1), _i = Si(16777216);
h = g.prototype;
h.ea = function() {
  if (j(this))
    return -m(this).ea();
  for (var t = 0, i = 1, n = 0; n < this.g.length; n++) {
    var e = this.D(n);
    t += (0 <= e ? e : Zt + e) * i, i *= Zt;
  }
  return t;
};
h.toString = function(t) {
  if (t = t || 10, 2 > t || 36 < t)
    throw Error("radix out of range: " + t);
  if (R(this))
    return "0";
  if (j(this))
    return "-" + m(this).toString(t);
  for (var i = I(Math.pow(t, 6)), n = this, e = ""; ; ) {
    var r = St(n, i).g;
    n = Et(n, r.R(i));
    var s = ((0 < n.g.length ? n.g[0] : n.h) >>> 0).toString(t);
    if (n = r, R(n))
      return s + e;
    for (; 6 > s.length; )
      s = "0" + s;
    e = s + e;
  }
};
h.D = function(t) {
  return 0 > t ? 0 : t < this.g.length ? this.g[t] : this.h;
};
function R(t) {
  if (t.h != 0)
    return !1;
  for (var i = 0; i < t.g.length; i++)
    if (t.g[i] != 0)
      return !1;
  return !0;
}
function j(t) {
  return t.h == -1;
}
h.X = function(t) {
  return t = Et(this, t), j(t) ? -1 : R(t) ? 0 : 1;
};
function m(t) {
  for (var i = t.g.length, n = [], e = 0; e < i; e++)
    n[e] = ~t.g[e];
  return new g(n, ~t.h).add(Qt);
}
h.abs = function() {
  return j(this) ? m(this) : this;
};
h.add = function(t) {
  for (var i = Math.max(this.g.length, t.g.length), n = [], e = 0, r = 0; r <= i; r++) {
    var s = e + (this.D(r) & 65535) + (t.D(r) & 65535), o = (s >>> 16) + (this.D(r) >>> 16) + (t.D(r) >>> 16);
    e = o >>> 16, s &= 65535, o &= 65535, n[r] = o << 16 | s;
  }
  return new g(n, n[n.length - 1] & -2147483648 ? -1 : 0);
};
function Et(t, i) {
  return t.add(m(i));
}
h.R = function(t) {
  if (R(this) || R(t))
    return B;
  if (j(this))
    return j(t) ? m(this).R(m(t)) : m(m(this).R(t));
  if (j(t))
    return m(this.R(m(t)));
  if (0 > this.X(_i) && 0 > t.X(_i))
    return I(this.ea() * t.ea());
  for (var i = this.g.length + t.g.length, n = [], e = 0; e < 2 * i; e++)
    n[e] = 0;
  for (e = 0; e < this.g.length; e++)
    for (var r = 0; r < t.g.length; r++) {
      var s = this.D(e) >>> 16, o = this.D(e) & 65535, f = t.D(r) >>> 16, a = t.D(r) & 65535;
      n[2 * e + 2 * r] += o * a, dt(n, 2 * e + 2 * r), n[2 * e + 2 * r + 1] += s * a, dt(n, 2 * e + 2 * r + 1), n[2 * e + 2 * r + 1] += o * f, dt(n, 2 * e + 2 * r + 1), n[2 * e + 2 * r + 2] += s * f, dt(n, 2 * e + 2 * r + 2);
    }
  for (e = 0; e < i; e++)
    n[e] = n[2 * e + 1] << 16 | n[2 * e];
  for (e = i; e < 2 * i; e++)
    n[e] = 0;
  return new g(n, 0);
};
function dt(t, i) {
  for (; (t[i] & 65535) != t[i]; )
    t[i + 1] += t[i] >>> 16, t[i] &= 65535, i++;
}
function Y(t, i) {
  this.g = t, this.h = i;
}
function St(t, i) {
  if (R(i))
    throw Error("division by zero");
  if (R(t))
    return new Y(B, B);
  if (j(t))
    return i = St(m(t), i), new Y(m(i.g), m(i.h));
  if (j(i))
    return i = St(t, m(i)), new Y(m(i.g), i.h);
  if (30 < t.g.length) {
    if (j(t) || j(i))
      throw Error("slowDivide_ only works with positive integers.");
    for (var n = Qt, e = i; 0 >= e.X(t); )
      n = Hi(n), e = Hi(e);
    var r = X(n, 1), s = X(e, 1);
    for (e = X(e, 2), n = X(n, 2); !R(e); ) {
      var o = s.add(e);
      0 >= o.X(t) && (r = r.add(n), s = o), e = X(e, 1), n = X(n, 1);
    }
    return i = Et(t, r.R(i)), new Y(r, i);
  }
  for (r = B; 0 <= t.X(i); ) {
    for (n = Math.max(1, Math.floor(t.ea() / i.ea())), e = Math.ceil(Math.log(n) / Math.LN2), e = 48 >= e ? 1 : Math.pow(2, e - 48), s = I(n), o = s.R(i); j(o) || 0 < o.X(t); )
      n -= e, s = I(n), o = s.R(i);
    R(s) && (s = Qt), r = r.add(s), t = Et(t, o);
  }
  return new Y(r, t);
}
h.gb = function(t) {
  return St(this, t).h;
};
h.and = function(t) {
  for (var i = Math.max(this.g.length, t.g.length), n = [], e = 0; e < i; e++)
    n[e] = this.D(e) & t.D(e);
  return new g(n, this.h & t.h);
};
h.or = function(t) {
  for (var i = Math.max(this.g.length, t.g.length), n = [], e = 0; e < i; e++)
    n[e] = this.D(e) | t.D(e);
  return new g(n, this.h | t.h);
};
h.xor = function(t) {
  for (var i = Math.max(this.g.length, t.g.length), n = [], e = 0; e < i; e++)
    n[e] = this.D(e) ^ t.D(e);
  return new g(n, this.h ^ t.h);
};
function Hi(t) {
  for (var i = t.g.length + 1, n = [], e = 0; e < i; e++)
    n[e] = t.D(e) << 1 | t.D(e - 1) >>> 31;
  return new g(n, t.h);
}
function X(t, i) {
  var n = i >> 5;
  i %= 32;
  for (var e = t.g.length - n, r = [], s = 0; s < e; s++)
    r[s] = 0 < i ? t.D(s + n) >>> i | t.D(s + n + 1) << 32 - i : t.D(s + n);
  return new g(r, t.h);
}
A.prototype.send = A.prototype.u;
A.prototype.open = A.prototype.m;
A.prototype.close = A.prototype.close;
pi.NO_ERROR = 0;
pi.TIMEOUT = 8;
pi.HTTP_ERROR = 6;
ce.COMPLETE = "complete";
ae.EventType = ht;
ht.OPEN = "a";
ht.CLOSE = "b";
ht.ERROR = "c";
ht.MESSAGE = "d";
v.prototype.listen = v.prototype.O;
c.prototype.listenOnce = c.prototype.P;
c.prototype.getLastError = c.prototype.Sa;
c.prototype.getLastErrorCode = c.prototype.Ia;
c.prototype.getStatus = c.prototype.da;
c.prototype.getResponseJson = c.prototype.Wa;
c.prototype.getResponseText = c.prototype.ja;
c.prototype.send = c.prototype.ha;
c.prototype.setWithCredentials = c.prototype.Oa;
M.prototype.digest = M.prototype.l;
M.prototype.reset = M.prototype.reset;
M.prototype.update = M.prototype.j;
g.prototype.add = g.prototype.add;
g.prototype.multiply = g.prototype.R;
g.prototype.modulo = g.prototype.gb;
g.prototype.compare = g.prototype.X;
g.prototype.toNumber = g.prototype.ea;
g.prototype.toString = g.prototype.toString;
g.prototype.getBits = g.prototype.D;
g.fromNumber = I;
g.fromString = Nn;
var Fe = g;
export {
  Fe as Integer
};
