// formik/dist/formik.esm.jsx
import ee from "../node_modules/deepmerge/dist/es.jsx";
import me from "../../lodash-es/isPlainObject.jsx";
import { r as c } from "../../react/index.jsx";
import L from "../../react-fast-compare/index.jsx";
import B from "../../tiny-warning/dist/tiny-warning.esm.jsx";
import ye from "../../lodash-es/clone.jsx";
import Te from "../../lodash-es/toPath.jsx";
import "../../hoist-non-react-statics/dist/hoist-non-react-statics.cjs.jsx";
function S() {
  return S = Object.assign || function(r) {
    for (var t = 1; t < arguments.length; t++) {
      var u = arguments[t];
      for (var i in u)
        Object.prototype.hasOwnProperty.call(u, i) && (r[i] = u[i]);
    }
    return r;
  }, S.apply(this, arguments);
}
function be(r, t) {
  if (r == null)
    return {};
  var u = {}, i = Object.keys(r), o, d;
  for (d = 0; d < i.length; d++)
    o = i[d], !(t.indexOf(o) >= 0) && (u[o] = r[o]);
  return u;
}
var j = /* @__PURE__ */ c.createContext(void 0);
j.displayName = "FormikContext";
j.Provider;
j.Consumer;
function He() {
  var r = c.useContext(j);
  return r || (process.env.NODE_ENV !== "production" ? B(!1, "Formik context is undefined, please verify you are calling useFormikContext() as child of a <Formik> component.") : B(!1)), r;
}
var R = function(t) {
  return typeof t == "function";
}, K = function(t) {
  return t !== null && typeof t == "object";
}, We = function(t) {
  return String(Math.floor(Number(t))) === t;
}, Z = function(t) {
  return Object.prototype.toString.call(t) === "[object String]";
}, x = function(t) {
  return K(t) && R(t.then);
};
function Ge(r) {
  if (r = r || (typeof document < "u" ? document : void 0), typeof r > "u")
    return null;
  try {
    return r.activeElement || r.body;
  } catch {
    return r.body;
  }
}
function T(r, t, u, i) {
  i === void 0 && (i = 0);
  for (var o = Te(t); r && i < o.length; )
    r = r[o[i++]];
  return i !== o.length && !r || r === void 0 ? u : r;
}
function M(r, t, u) {
  for (var i = ye(r), o = i, d = 0, v = Te(t); d < v.length - 1; d++) {
    var m = v[d], _ = T(r, v.slice(0, d + 1));
    if (_ && (K(_) || Array.isArray(_)))
      o = o[m] = ye(_);
    else {
      var g = v[d + 1];
      o = o[m] = We(g) && Number(g) >= 0 ? [] : {};
    }
  }
  return (d === 0 ? r : o)[v[d]] === u ? r : (u === void 0 ? delete o[v[d]] : o[v[d]] = u, d === 0 && u === void 0 && delete i[v[d]], i);
}
function ge(r, t, u, i) {
  u === void 0 && (u = /* @__PURE__ */ new WeakMap()), i === void 0 && (i = {});
  for (var o = 0, d = Object.keys(r); o < d.length; o++) {
    var v = d[o], m = r[v];
    K(m) ? u.get(m) || (u.set(m, !0), i[v] = Array.isArray(m) ? [] : {}, ge(m, t, u, i[v])) : i[v] = t;
  }
  return i;
}
function je(r, t) {
  switch (t.type) {
    case "SET_VALUES":
      return S({}, r, {
        values: t.payload
      });
    case "SET_TOUCHED":
      return S({}, r, {
        touched: t.payload
      });
    case "SET_ERRORS":
      return L(r.errors, t.payload) ? r : S({}, r, {
        errors: t.payload
      });
    case "SET_STATUS":
      return S({}, r, {
        status: t.payload
      });
    case "SET_ISSUBMITTING":
      return S({}, r, {
        isSubmitting: t.payload
      });
    case "SET_ISVALIDATING":
      return S({}, r, {
        isValidating: t.payload
      });
    case "SET_FIELD_VALUE":
      return S({}, r, {
        values: M(r.values, t.payload.field, t.payload.value)
      });
    case "SET_FIELD_TOUCHED":
      return S({}, r, {
        touched: M(r.touched, t.payload.field, t.payload.value)
      });
    case "SET_FIELD_ERROR":
      return S({}, r, {
        errors: M(r.errors, t.payload.field, t.payload.value)
      });
    case "RESET_FORM":
      return S({}, r, t.payload);
    case "SET_FORMIK_STATE":
      return t.payload(r);
    case "SUBMIT_ATTEMPT":
      return S({}, r, {
        touched: ge(r.values, !0),
        isSubmitting: !0,
        submitCount: r.submitCount + 1
      });
    case "SUBMIT_FAILURE":
      return S({}, r, {
        isSubmitting: !1
      });
    case "SUBMIT_SUCCESS":
      return S({}, r, {
        isSubmitting: !1
      });
    default:
      return r;
  }
}
var D = {}, G = {};
function ar(r) {
  var t = r.validateOnChange, u = t === void 0 ? !0 : t, i = r.validateOnBlur, o = i === void 0 ? !0 : i, d = r.validateOnMount, v = d === void 0 ? !1 : d, m = r.isInitialValid, _ = r.enableReinitialize, g = _ === void 0 ? !1 : _, te = r.onSubmit, Fe = be(r, ["validateOnChange", "validateOnBlur", "validateOnMount", "isInitialValid", "enableReinitialize", "onSubmit"]), s = S({
    validateOnChange: u,
    validateOnBlur: o,
    validateOnMount: v,
    onSubmit: te
  }, Fe), F = c.useRef(s.initialValues), O = c.useRef(s.initialErrors || D), k = c.useRef(s.initialTouched || G), N = c.useRef(s.initialStatus), I = c.useRef(!1), V = c.useRef({});
  process.env.NODE_ENV !== "production" && c.useEffect(function() {
    typeof m > "u" || (process.env.NODE_ENV !== "production" ? B(!1, "isInitialValid has been deprecated and will be removed in future versions of Formik. Please use initialErrors or validateOnMount instead.") : B(!1));
  }, []), c.useEffect(function() {
    return I.current = !0, function() {
      I.current = !1;
    };
  }, []);
  var Ie = c.useState(0), Ae = Ie[1], H = c.useRef({
    values: s.initialValues,
    errors: s.initialErrors || D,
    touched: s.initialTouched || G,
    status: s.initialStatus,
    isSubmitting: !1,
    isValidating: !1,
    submitCount: 0
  }), E = H.current, h = c.useCallback(function(e) {
    var n = H.current;
    H.current = je(n, e), n !== H.current && Ae(function(a) {
      return a + 1;
    });
  }, []), ne = c.useCallback(function(e, n) {
    return new Promise(function(a, l) {
      var f = s.validate(e, n);
      f == null ? a(D) : x(f) ? f.then(function(p) {
        a(p || D);
      }, function(p) {
        process.env.NODE_ENV !== "production" && console.warn("Warning: An unhandled error was caught during validation in <Formik validate />", p), l(p);
      }) : a(f);
    });
  }, [s.validate]), $ = c.useCallback(function(e, n) {
    var a = s.validationSchema, l = R(a) ? a(n) : a, f = n && l.validateAt ? l.validateAt(n, e) : $e(e, l);
    return new Promise(function(p, y) {
      f.then(function() {
        p(D);
      }, function(C) {
        C.name === "ValidationError" ? p(Ke(C)) : (process.env.NODE_ENV !== "production" && console.warn("Warning: An unhandled error was caught during validation in <Formik validationSchema />", C), y(C));
      });
    });
  }, [s.validationSchema]), ie = c.useCallback(function(e, n) {
    return new Promise(function(a) {
      return a(V.current[e].validate(n));
    });
  }, []), ae = c.useCallback(function(e) {
    var n = Object.keys(V.current).filter(function(l) {
      return R(V.current[l].validate);
    }), a = n.length > 0 ? n.map(function(l) {
      return ie(l, T(e, l));
    }) : [Promise.resolve("DO_NOT_DELETE_YOU_WILL_BE_FIRED")];
    return Promise.all(a).then(function(l) {
      return l.reduce(function(f, p, y) {
        return p === "DO_NOT_DELETE_YOU_WILL_BE_FIRED" || p && (f = M(f, n[y], p)), f;
      }, {});
    });
  }, [ie]), Re = c.useCallback(function(e) {
    return Promise.all([ae(e), s.validationSchema ? $(e) : {}, s.validate ? ne(e) : {}]).then(function(n) {
      var a = n[0], l = n[1], f = n[2], p = ee.all([a, l, f], {
        arrayMerge: ze
      });
      return p;
    });
  }, [s.validate, s.validationSchema, ae, ne, $]), A = b(function(e) {
    return e === void 0 && (e = E.values), h({
      type: "SET_ISVALIDATING",
      payload: !0
    }), Re(e).then(function(n) {
      return I.current && (h({
        type: "SET_ISVALIDATING",
        payload: !1
      }), h({
        type: "SET_ERRORS",
        payload: n
      })), n;
    });
  });
  c.useEffect(function() {
    v && I.current === !0 && L(F.current, s.initialValues) && A(F.current);
  }, [v, A]);
  var P = c.useCallback(function(e) {
    var n = e && e.values ? e.values : F.current, a = e && e.errors ? e.errors : O.current ? O.current : s.initialErrors || {}, l = e && e.touched ? e.touched : k.current ? k.current : s.initialTouched || {}, f = e && e.status ? e.status : N.current ? N.current : s.initialStatus;
    F.current = n, O.current = a, k.current = l, N.current = f;
    var p = function() {
      h({
        type: "RESET_FORM",
        payload: {
          isSubmitting: !!e && !!e.isSubmitting,
          errors: a,
          touched: l,
          status: f,
          values: n,
          isValidating: !!e && !!e.isValidating,
          submitCount: e && e.submitCount && typeof e.submitCount == "number" ? e.submitCount : 0
        }
      });
    };
    if (s.onReset) {
      var y = s.onReset(E.values, he);
      x(y) ? y.then(p) : p();
    } else
      p();
  }, [s.initialErrors, s.initialStatus, s.initialTouched, s.onReset]);
  c.useEffect(function() {
    I.current === !0 && !L(F.current, s.initialValues) && g && (F.current = s.initialValues, P(), v && A(F.current));
  }, [g, s.initialValues, P, v, A]), c.useEffect(function() {
    g && I.current === !0 && !L(O.current, s.initialErrors) && (O.current = s.initialErrors || D, h({
      type: "SET_ERRORS",
      payload: s.initialErrors || D
    }));
  }, [g, s.initialErrors]), c.useEffect(function() {
    g && I.current === !0 && !L(k.current, s.initialTouched) && (k.current = s.initialTouched || G, h({
      type: "SET_TOUCHED",
      payload: s.initialTouched || G
    }));
  }, [g, s.initialTouched]), c.useEffect(function() {
    g && I.current === !0 && !L(N.current, s.initialStatus) && (N.current = s.initialStatus, h({
      type: "SET_STATUS",
      payload: s.initialStatus
    }));
  }, [g, s.initialStatus, s.initialTouched]);
  var ue = b(function(e) {
    if (V.current[e] && R(V.current[e].validate)) {
      var n = T(E.values, e), a = V.current[e].validate(n);
      return x(a) ? (h({
        type: "SET_ISVALIDATING",
        payload: !0
      }), a.then(function(l) {
        return l;
      }).then(function(l) {
        h({
          type: "SET_FIELD_ERROR",
          payload: {
            field: e,
            value: l
          }
        }), h({
          type: "SET_ISVALIDATING",
          payload: !1
        });
      })) : (h({
        type: "SET_FIELD_ERROR",
        payload: {
          field: e,
          value: a
        }
      }), Promise.resolve(a));
    } else if (s.validationSchema)
      return h({
        type: "SET_ISVALIDATING",
        payload: !0
      }), $(E.values, e).then(function(l) {
        return l;
      }).then(function(l) {
        h({
          type: "SET_FIELD_ERROR",
          payload: {
            field: e,
            value: T(l, e)
          }
        }), h({
          type: "SET_ISVALIDATING",
          payload: !1
        });
      });
    return Promise.resolve();
  }), Ce = c.useCallback(function(e, n) {
    var a = n.validate;
    V.current[e] = {
      validate: a
    };
  }, []), _e = c.useCallback(function(e) {
    delete V.current[e];
  }, []), oe = b(function(e, n) {
    h({
      type: "SET_TOUCHED",
      payload: e
    });
    var a = n === void 0 ? o : n;
    return a ? A(E.values) : Promise.resolve();
  }), le = c.useCallback(function(e) {
    h({
      type: "SET_ERRORS",
      payload: e
    });
  }, []), se = b(function(e, n) {
    var a = R(e) ? e(E.values) : e;
    h({
      type: "SET_VALUES",
      payload: a
    });
    var l = n === void 0 ? u : n;
    return l ? A(a) : Promise.resolve();
  }), W = c.useCallback(function(e, n) {
    h({
      type: "SET_FIELD_ERROR",
      payload: {
        field: e,
        value: n
      }
    });
  }, []), w = b(function(e, n, a) {
    h({
      type: "SET_FIELD_VALUE",
      payload: {
        field: e,
        value: n
      }
    });
    var l = a === void 0 ? u : a;
    return l ? A(M(E.values, e, n)) : Promise.resolve();
  }), ce = c.useCallback(function(e, n) {
    var a = n, l = e, f;
    if (!Z(e)) {
      e.persist && e.persist();
      var p = e.target ? e.target : e.currentTarget, y = p.type, C = p.name, Q = p.id, X = p.value, Ue = p.checked, Be = p.outerHTML, Ee = p.options, Pe = p.multiple;
      a = n || C || Q, !a && process.env.NODE_ENV !== "production" && Se({
        htmlContent: Be,
        documentationAnchorLink: "handlechange-e-reactchangeeventany--void",
        handlerName: "handleChange"
      }), l = /number|range/.test(y) ? (f = parseFloat(X), isNaN(f) ? "" : f) : /checkbox/.test(y) ? qe(T(E.values, a), Ue, X) : Ee && Pe ? Ye(Ee) : X;
    }
    a && w(a, l);
  }, [w, E.values]), z = b(function(e) {
    if (Z(e))
      return function(n) {
        return ce(n, e);
      };
    ce(e);
  }), U = b(function(e, n, a) {
    n === void 0 && (n = !0), h({
      type: "SET_FIELD_TOUCHED",
      payload: {
        field: e,
        value: n
      }
    });
    var l = a === void 0 ? o : a;
    return l ? A(E.values) : Promise.resolve();
  }), de = c.useCallback(function(e, n) {
    e.persist && e.persist();
    var a = e.target, l = a.name, f = a.id, p = a.outerHTML, y = n || l || f;
    !y && process.env.NODE_ENV !== "production" && Se({
      htmlContent: p,
      documentationAnchorLink: "handleblur-e-any--void",
      handlerName: "handleBlur"
    }), U(y, !0);
  }, [U]), Y = b(function(e) {
    if (Z(e))
      return function(n) {
        return de(n, e);
      };
    de(e);
  }), fe = c.useCallback(function(e) {
    R(e) ? h({
      type: "SET_FORMIK_STATE",
      payload: e
    }) : h({
      type: "SET_FORMIK_STATE",
      payload: function() {
        return e;
      }
    });
  }, []), ve = c.useCallback(function(e) {
    h({
      type: "SET_STATUS",
      payload: e
    });
  }, []), pe = c.useCallback(function(e) {
    h({
      type: "SET_ISSUBMITTING",
      payload: e
    });
  }, []), q = b(function() {
    return h({
      type: "SUBMIT_ATTEMPT"
    }), A().then(function(e) {
      var n = e instanceof Error, a = !n && Object.keys(e).length === 0;
      if (a) {
        var l;
        try {
          if (l = Oe(), l === void 0)
            return;
        } catch (f) {
          throw f;
        }
        return Promise.resolve(l).then(function(f) {
          return I.current && h({
            type: "SUBMIT_SUCCESS"
          }), f;
        }).catch(function(f) {
          if (I.current)
            throw h({
              type: "SUBMIT_FAILURE"
            }), f;
        });
      } else if (I.current && (h({
        type: "SUBMIT_FAILURE"
      }), n))
        throw e;
    });
  }), Ve = b(function(e) {
    if (e && e.preventDefault && R(e.preventDefault) && e.preventDefault(), e && e.stopPropagation && R(e.stopPropagation) && e.stopPropagation(), process.env.NODE_ENV !== "production" && typeof document < "u") {
      var n = Ge();
      n !== null && n instanceof HTMLButtonElement && (n.attributes && n.attributes.getNamedItem("type") || (process.env.NODE_ENV !== "production" ? B(!1, 'You submitted a Formik form using a button with an unspecified `type` attribute.  Most browsers default button elements to `type="submit"`. If this is not a submit button, please add `type="button"`.') : B(!1)));
    }
    q().catch(function(a) {
      console.warn("Warning: An unhandled error was caught from submitForm()", a);
    });
  }), he = {
    resetForm: P,
    validateForm: A,
    validateField: ue,
    setErrors: le,
    setFieldError: W,
    setFieldTouched: U,
    setFieldValue: w,
    setStatus: ve,
    setSubmitting: pe,
    setTouched: oe,
    setValues: se,
    setFormikState: fe,
    submitForm: q
  }, Oe = b(function() {
    return te(E.values, he);
  }), ke = b(function(e) {
    e && e.preventDefault && R(e.preventDefault) && e.preventDefault(), e && e.stopPropagation && R(e.stopPropagation) && e.stopPropagation(), P();
  }), De = c.useCallback(function(e) {
    return {
      value: T(E.values, e),
      error: T(E.errors, e),
      touched: !!T(E.touched, e),
      initialValue: T(F.current, e),
      initialTouched: !!T(k.current, e),
      initialError: T(O.current, e)
    };
  }, [E.errors, E.touched, E.values]), Le = c.useCallback(function(e) {
    return {
      setValue: function(a, l) {
        return w(e, a, l);
      },
      setTouched: function(a, l) {
        return U(e, a, l);
      },
      setError: function(a) {
        return W(e, a);
      }
    };
  }, [w, U, W]), Me = c.useCallback(function(e) {
    var n = K(e), a = n ? e.name : e, l = T(E.values, a), f = {
      name: a,
      value: l,
      onChange: z,
      onBlur: Y
    };
    if (n) {
      var p = e.type, y = e.value, C = e.as, Q = e.multiple;
      p === "checkbox" ? y === void 0 ? f.checked = !!l : (f.checked = !!(Array.isArray(l) && ~l.indexOf(y)), f.value = y) : p === "radio" ? (f.checked = l === y, f.value = y) : C === "select" && Q && (f.value = f.value || [], f.multiple = !0);
    }
    return f;
  }, [Y, z, E.values]), J = c.useMemo(function() {
    return !L(F.current, E.values);
  }, [F.current, E.values]), Ne = c.useMemo(function() {
    return typeof m < "u" ? J ? E.errors && Object.keys(E.errors).length === 0 : m !== !1 && R(m) ? m(s) : m : E.errors && Object.keys(E.errors).length === 0;
  }, [m, J, E.errors, s]), we = S({}, E, {
    initialValues: F.current,
    initialErrors: O.current,
    initialTouched: k.current,
    initialStatus: N.current,
    handleBlur: Y,
    handleChange: z,
    handleReset: ke,
    handleSubmit: Ve,
    resetForm: P,
    setErrors: le,
    setFormikState: fe,
    setFieldTouched: U,
    setFieldValue: w,
    setFieldError: W,
    setStatus: ve,
    setSubmitting: pe,
    setTouched: oe,
    setValues: se,
    submitForm: q,
    validateForm: A,
    validateField: ue,
    isValid: Ne,
    dirty: J,
    unregisterField: _e,
    registerField: Ce,
    getFieldProps: Me,
    getFieldMeta: De,
    getFieldHelpers: Le,
    validateOnBlur: o,
    validateOnChange: u,
    validateOnMount: v
  });
  return we;
}
function Se(r) {
  var t = r.htmlContent, u = r.documentationAnchorLink, i = r.handlerName;
  console.warn("Warning: Formik called `" + i + "`, but you forgot to pass an `id` or `name` attribute to your input:\n    " + t + `
    Formik cannot determine which value to update. For more info see https://formik.org/docs/api/formik#` + u + `
  `);
}
function Ke(r) {
  var t = {};
  if (r.inner) {
    if (r.inner.length === 0)
      return M(t, r.path, r.message);
    for (var o = r.inner, u = Array.isArray(o), i = 0, o = u ? o : o[Symbol.iterator](); ; ) {
      var d;
      if (u) {
        if (i >= o.length)
          break;
        d = o[i++];
      } else {
        if (i = o.next(), i.done)
          break;
        d = i.value;
      }
      var v = d;
      T(t, v.path) || (t = M(t, v.path, v.message));
    }
  }
  return t;
}
function $e(r, t, u, i) {
  u === void 0 && (u = !1);
  var o = re(r);
  return t[u ? "validateSync" : "validate"](o, {
    abortEarly: !1,
    context: i || o
  });
}
function re(r) {
  var t = Array.isArray(r) ? [] : {};
  for (var u in r)
    if (Object.prototype.hasOwnProperty.call(r, u)) {
      var i = String(u);
      Array.isArray(r[i]) === !0 ? t[i] = r[i].map(function(o) {
        return Array.isArray(o) === !0 || me(o) ? re(o) : o !== "" ? o : void 0;
      }) : me(r[i]) ? t[i] = re(r[i]) : t[i] = r[i] !== "" ? r[i] : void 0;
    }
  return t;
}
function ze(r, t, u) {
  var i = r.slice();
  return t.forEach(function(d, v) {
    if (typeof i[v] > "u") {
      var m = u.clone !== !1, _ = m && u.isMergeableObject(d);
      i[v] = _ ? ee(Array.isArray(d) ? [] : {}, d, u) : d;
    } else
      u.isMergeableObject(d) ? i[v] = ee(r[v], d, u) : r.indexOf(d) === -1 && i.push(d);
  }), i;
}
function Ye(r) {
  return Array.from(r).filter(function(t) {
    return t.selected;
  }).map(function(t) {
    return t.value;
  });
}
function qe(r, t, u) {
  if (typeof r == "boolean")
    return !!t;
  var i = [], o = !1, d = -1;
  if (Array.isArray(r))
    i = r, d = r.indexOf(u), o = d >= 0;
  else if (!u || u == "true" || u == "false")
    return !!t;
  return t && u && !o ? i.concat(u) : o ? i.slice(0, d).concat(i.slice(d + 1)) : i;
}
var Je = typeof window < "u" && typeof window.document < "u" && typeof window.document.createElement < "u" ? c.useLayoutEffect : c.useEffect;
function b(r) {
  var t = c.useRef(r);
  return Je(function() {
    t.current = r;
  }), c.useCallback(function() {
    for (var u = arguments.length, i = new Array(u), o = 0; o < u; o++)
      i[o] = arguments[o];
    return t.current.apply(void 0, i);
  }, []);
}
var Qe = /* @__PURE__ */ c.forwardRef(function(r, t) {
  var u = r.action, i = be(r, ["action"]), o = u ?? "#", d = He(), v = d.handleReset, m = d.handleSubmit;
  return c.createElement("form", S({
    onSubmit: m,
    ref: t,
    onReset: v,
    action: o
  }, i));
});
Qe.displayName = "Form";
export {
  Qe as Form,
  j as FormikContext,
  Ge as getActiveElement,
  T as getIn,
  R as isFunction,
  We as isInteger,
  K as isObject,
  x as isPromise,
  Z as isString,
  re as prepareDataForValidation,
  M as setIn,
  ge as setNestedObjectValues,
  ar as useFormik,
  He as useFormikContext,
  $e as validateYupSchema,
  Ke as yupToFormErrors
};
