// @tanstack/react-table/build/lib/index.js
import { r as a } from "../../../../react/index.js";
import { createTable as s } from "../../../table-core/build/lib/index.js";
import { ColumnSizing as m, Expanding as p, Filters as R, Grouping as C, Headers as f, Ordering as w, Pagination as b, Pinning as F, RowSelection as h, Sorting as x, Visibility as v, aggregationFns as y, buildHeaderGroups as A, createCell as E, createColumn as M, createRow as O, defaultColumnSizing as P, expandRows as k, filterFns as z, flattenBy as G, functionalUpdate as H, getCoreRowModel as N, getPaginationRowModel as T, getSortedRowModel as U, isFunction as V, isNumberArray as B, isRowSelected as j, isSubRowSelected as q, makeStateUpdater as D, memo as I, orderColumns as J, passiveEventSupported as K, reSplitAlphaNumeric as L, selectRowsFn as Q, shouldAutoRemoveFilter as W, sortingFns as X } from "../../../table-core/build/lib/index.js";
/**
   * react-table
   *
   * Copyright (c) TanStack
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE.md file in the root directory of this source tree.
   *
   * @license MIT
   */
function g(e) {
  const r = {
    state: {},
    // Dummy state
    onStateChange: () => {
    },
    // noop
    renderFallbackValue: null,
    ...e
  }, [t] = a.useState(() => ({
    current: s(r)
  })), [o, i] = a.useState(() => t.current.initialState);
  return t.current.setOptions((l) => ({
    ...l,
    ...e,
    state: {
      ...o,
      ...e.state
    },
    // Similarly, we'll maintain both our internal state and any user-provided
    // state.
    onStateChange: (n) => {
      i(n), e.onStateChange == null || e.onStateChange(n);
    }
  })), t.current;
}
export {
  m as ColumnSizing,
  p as Expanding,
  R as Filters,
  C as Grouping,
  f as Headers,
  w as Ordering,
  b as Pagination,
  F as Pinning,
  h as RowSelection,
  x as Sorting,
  v as Visibility,
  y as aggregationFns,
  A as buildHeaderGroups,
  E as createCell,
  M as createColumn,
  O as createRow,
  s as createTable,
  P as defaultColumnSizing,
  k as expandRows,
  z as filterFns,
  G as flattenBy,
  H as functionalUpdate,
  N as getCoreRowModel,
  T as getPaginationRowModel,
  U as getSortedRowModel,
  V as isFunction,
  B as isNumberArray,
  j as isRowSelected,
  q as isSubRowSelected,
  D as makeStateUpdater,
  I as memo,
  J as orderColumns,
  K as passiveEventSupported,
  L as reSplitAlphaNumeric,
  Q as selectRowsFn,
  W as shouldAutoRemoveFilter,
  X as sortingFns,
  g as useReactTable
};
