// firebase/app/dist/esm/index.esm.js
import { registerVersion as e } from "../../../../@firebase/app/dist/esm/index.esm2017.js";
import { SDK_VERSION as n, _DEFAULT_ENTRY_NAME as _, _addComponent as a, _apps as m, _components as s, _getProvider as g, _registerComponent as v, getApp as A, initializeApp as E } from "../../../../@firebase/app/dist/esm/index.esm2017.js";
var r = "firebase", o = "10.7.2";
/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
e(r, o, "app");
export {
  n as SDK_VERSION,
  _ as _DEFAULT_ENTRY_NAME,
  a as _addComponent,
  m as _apps,
  s as _components,
  g as _getProvider,
  v as _registerComponent,
  A as getApp,
  E as initializeApp,
  e as registerVersion
};
